// @kepler.gl/actions/dist/index.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  VisStateActions: true,
  MapStateActions: true,
  UIStateActions: true,
  MapStyleActions: true,
  ProviderActionTypes: true,
  ProviderActions: true,
  _actionFor: true,
  forwardTo: true,
  getActionForwardAddress: true,
  isForwardAction: true,
  unwrap: true,
  wrapTo: true,
  ACTION_PREFIX: true,
  ActionTypes: true
};
Object.defineProperty(exports, "ProviderActionTypes", {
  enumerable: true,
  get: function get() {
    return ProviderActions.ActionTypes;
  }
});
Object.defineProperty(exports, "_actionFor", {
  enumerable: true,
  get: function get() {
    return _actionWrapper._actionFor;
  }
});
Object.defineProperty(exports, "forwardTo", {
  enumerable: true,
  get: function get() {
    return _actionWrapper.forwardTo;
  }
});
Object.defineProperty(exports, "getActionForwardAddress", {
  enumerable: true,
  get: function get() {
    return _actionWrapper.getActionForwardAddress;
  }
});
Object.defineProperty(exports, "isForwardAction", {
  enumerable: true,
  get: function get() {
    return _actionWrapper.isForwardAction;
  }
});
Object.defineProperty(exports, "unwrap", {
  enumerable: true,
  get: function get() {
    return _actionWrapper.unwrap;
  }
});
Object.defineProperty(exports, "wrapTo", {
  enumerable: true,
  get: function get() {
    return _actionWrapper.wrapTo;
  }
});
Object.defineProperty(exports, "ACTION_PREFIX", {
  enumerable: true,
  get: function get() {
    return _actionTypes.ACTION_PREFIX;
  }
});
Object.defineProperty(exports, "ActionTypes", {
  enumerable: true,
  get: function get() {
    return _actionTypes["default"];
  }
});
exports.ProviderActions = exports.MapStyleActions = exports.UIStateActions = exports.MapStateActions = exports.VisStateActions = void 0;

var VisStateActions = _interopRequireWildcard(require("./vis-state-actions"));

exports.VisStateActions = VisStateActions;
Object.keys(VisStateActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === VisStateActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return VisStateActions[key];
    }
  });
});

var MapStateActions = _interopRequireWildcard(require("./map-state-actions"));

exports.MapStateActions = MapStateActions;
Object.keys(MapStateActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === MapStateActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return MapStateActions[key];
    }
  });
});

var UIStateActions = _interopRequireWildcard(require("./ui-state-actions"));

exports.UIStateActions = UIStateActions;
Object.keys(UIStateActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === UIStateActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return UIStateActions[key];
    }
  });
});

var MapStyleActions = _interopRequireWildcard(require("./map-style-actions"));

exports.MapStyleActions = MapStyleActions;
Object.keys(MapStyleActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === MapStyleActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return MapStyleActions[key];
    }
  });
});

var ProviderActions = _interopRequireWildcard(require("./provider-actions"));

exports.ProviderActions = ProviderActions;
Object.keys(ProviderActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === ProviderActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return ProviderActions[key];
    }
  });
});

var _actions = require("./actions");

Object.keys(_actions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _actions[key];
    }
  });
});

var _actionWrapper = require("./action-wrapper");

Object.keys(_actionWrapper).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actionWrapper[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _actionWrapper[key];
    }
  });
});

var _identityActions = require("./identity-actions");

Object.keys(_identityActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _identityActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _identityActions[key];
    }
  });
});

var _actionTypes = _interopRequireWildcard(require("./action-types"));
