// react-map-gl-draw/dist-es6/editor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _mjolnir = require("mjolnir.js");

var _reactMapGl = require("react-map-gl");

var _uuid = _interopRequireDefault(require("uuid"));

var _feature2 = _interopRequireDefault(require("./feature"));

var _style = require("./style");

var _constants = require("./constants");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultProps = {
  mode: _constants.MODES.READ_ONLY,
  clickRadius: 0,
  getEditHandleStyle: _style.getEditHandleStyle,
  getFeatureStyle: _style.getFeatureStyle,
  getFeatureShape: 'circle',
  getEditHandleShape: 'circle',
  onSelect: function onSelect() {}
};
var UNCOMMITTED_ID = 'uncommitted';

var Editor =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Editor, _PureComponent);

  function Editor(props) {
    var _this;

    _classCallCheck(this, Editor);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Editor).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_containerRef", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_events", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_context", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_update", function (features) {
      if (features) {
        _this.props.onUpdate(features.map(function (f) {
          return f.toFeature();
        }));
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_updateRectangle", function (feature, options) {
      var vertexIndex = options.vertexIndex,
          lngLat = options.lngLat;
      /*
      *   p0.x, p0.y   ----------  diagonal.x, p0.y
      *       |                             |
      *       |                             |
      *   p0.x, diagonal.y ----- diagonal.x, diagonal.y
      */

      var diagonal = vertexIndex;
      var p0 = feature.points[(diagonal + 2) % 4];
      feature.replacePoint(diagonal, [lngLat[0], lngLat[1]]);
      feature.replacePoint((diagonal + 1) % 4, [lngLat[0], p0[1]]);
      feature.replacePoint((diagonal + 3) % 4, [p0[0], lngLat[1]]);

      _this._update(_this.state.features);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_updateFeature", function (feature, mode, options) {
      switch (mode) {
        case 'vertex':
          if (feature.renderType === _constants.RENDER_TYPE.RECTANGLE) {
            _this._updateRectangle(feature, options);
          } else {
            feature.replacePoint(options.vertexIndex, [options.lngLat[0], options.lngLat[1]]);

            _this._update(_this.state.features);
          }

          break;

        case 'feature':
          var dx = options.dx,
              dy = options.dy;
          feature.points = feature.points.map(function (lngLat) {
            var pixels = _this._project(lngLat);

            if (pixels) {
              pixels[0] += dx;
              pixels[1] += dy;
              return _this._unproject(pixels);
            }

            return null;
          }).filter(Boolean);

          _this._update(_this.state.features);

          break;

        case 'Rectangle':
          _this._updateRectangle(feature, options);

          break;

        default:
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_addPoint", function (x, y, feature) {
      var isNew = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
      feature = feature || _this._getSelectedFeature();

      if (!feature) {
        return;
      }

      var lngLat = _this._unproject([x, y]);

      if (!lngLat) {
        return;
      }

      feature.addPoint([lngLat[0], lngLat[1]]);
      var features = _this.state.features || [];

      if (isNew) {
        features.push(feature);
      }

      var validPath = feature.points && feature.points.length >= 2;
      var _this$props = _this.props,
          mode = _this$props.mode,
          onSelect = _this$props.onSelect;

      if (mode === _constants.MODES.DRAW_POINT || mode === _constants.MODES.DRAW_PATH && validPath) {
        _this._update(features);

        onSelect({
          selectedFeatureId: feature && feature.id
        });
      } else {
        _this.setState({
          features: _toConsumableArray(features),
          selectedFeatureId: feature && feature.id
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_clearCache", function () {
      _this.setState({
        selectedFeatureId: null,
        uncommittedLngLat: null,
        hoveredFeatureId: null,
        hoveredLngLat: null,
        hoveredVertexIndex: -1,
        draggingVertexIndex: -1,
        startDragPos: null,
        isDragging: false,
        didDrag: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_closePath", function () {
      var selectedFeature = _this._getSelectedFeature();

      if (selectedFeature) {
        selectedFeature.closePath();

        _this._update(_this.state.features);
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_addFeature", function (mode, point) {
      var type = _constants.MODE_TO_GEOJSON_TYPE[mode];
      var renderType = _constants.MODE_TO_RENDER_TYPE[mode];
      var feature = new _feature2.default({
        id: (0, _uuid.default)(),
        type: type,
        renderType: renderType
      });

      _this._addPoint(point.x, point.y, feature, true);

      if (mode === _constants.MODES.DRAW_RECTANGLE) {
        for (var i = 0; i < 3; i++) {
          _this._addPoint(point.x, point.y, feature, false);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onEvent", function (handler, evt, stopPropagation) {
      var mode = _this.props.mode;

      if (mode === _constants.MODES.READ_ONLY) {
        return;
      }

      for (var _len = arguments.length, args = new Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
        args[_key - 3] = arguments[_key];
      }

      handler.apply(void 0, [evt].concat(args));

      if (stopPropagation) {
        evt.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onMouseUp", function (evt) {
      _this.setState({
        isDragging: false,
        didDrag: false
      });

      var draggingVertexIndex = _this.state.draggingVertexIndex;

      if (Number(draggingVertexIndex) >= 0) {
        _this.setState({
          draggingVertexIndex: -1
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onMouseDown", function (evt) {
      var _this$_getEventPositi = _this._getEventPosition(evt),
          x = _this$_getEventPositi.x,
          y = _this$_getEventPositi.y;

      var attributes = (0, _utils.parseElemDataAttributes)(evt.target); // click segment

      if (attributes && attributes.type === _constants.ELEMENT_TYPE.VERTEX) {
        var vertexIndex = attributes.vertexIndex;

        _this.setState({
          draggingVertexIndex: vertexIndex,
          startDragPos: {
            x: x,
            y: y
          },
          isDragging: true
        }); // click selected feature

      } else if (_this._matchesFeature(attributes, _this._getSelectedFeature())) {
        _this.setState({
          startDragPos: {
            x: x,
            y: y
          },
          isDragging: true
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onMouseMove", function (evt) {
      var attributes = (0, _utils.parseElemDataAttributes)(evt.target) || {};
      var vertexIndex = attributes.vertexIndex,
          featureIndex = attributes.featureIndex,
          type = attributes.type;
      var _this$state = _this.state,
          startDragPos = _this$state.startDragPos,
          isDragging = _this$state.isDragging,
          didDrag = _this$state.didDrag;
      var mode = _this.props.mode;

      var _this$_getEventPositi2 = _this._getEventPosition(evt),
          x = _this$_getEventPositi2.x,
          y = _this$_getEventPositi2.y;

      var lngLat = _this._unproject([x, y]);

      if (isDragging && !didDrag && startDragPos) {
        var dx = x - startDragPos.x;
        var dy = y - startDragPos.y;

        if (dx * dx + dy * dy > 5) {
          _this.setState({
            didDrag: true
          });
        }
      }

      var selectedFeature = _this._getSelectedFeature();

      var isDrawing = _constants.DRAWING_MODES.indexOf(mode) !== -1;
      var isEditing = mode === _constants.MODES.EDIT_VERTEX;

      if (selectedFeature) {
        // dragging
        if (didDrag && startDragPos) {
          var draggingVertexIndex = Number(_this.state.draggingVertexIndex);

          if (draggingVertexIndex >= 0) {
            // dragging vertex
            _this._updateFeature(selectedFeature, 'vertex', {
              vertexIndex: draggingVertexIndex,
              lngLat: lngLat
            });
          } else {
            // dragging feature
            var _dx = x - startDragPos.x;

            var _dy = y - startDragPos.y;

            _this.setState({
              startDragPos: {
                x: x,
                y: y
              }
            });

            _this._updateFeature(selectedFeature, 'feature', {
              dx: _dx,
              dy: _dy
            });
          }
        } else if (mode === _constants.MODES.DRAW_RECTANGLE) {
          // drawing rectangle
          _this._updateFeature(selectedFeature, 'Rectangle', {
            vertexIndex: 2,
            lngLat: lngLat
          });
        } else if (isDrawing) {
          // drawing other shapes
          _this.setState({
            uncommittedLngLat: lngLat
          });
        } else if (isEditing) {
          if ((selectedFeature.renderType === _constants.RENDER_TYPE.LINE_STRING || selectedFeature.renderType === _constants.RENDER_TYPE.POLYGON) && type === _constants.ELEMENT_TYPE.SEGMENT) {
            // segmentId is start vertexIndex
            var uncommittedLngLat = null;

            if (lngLat && typeof vertexIndex === 'number') {
              uncommittedLngLat = _this._getClosestPositionOnSegment(vertexIndex, lngLat, selectedFeature);
            }

            _this.setState({
              uncommittedLngLat: uncommittedLngLat
            });
          } else {
            _this.setState({
              uncommittedLngLat: null
            });
          }
        }
      }

      var _this$state2 = _this.state,
          features = _this$state2.features,
          selectedFeatureId = _this$state2.selectedFeatureId;

      if (selectedFeatureId && type === _constants.ELEMENT_TYPE.VERTEX && typeof featureIndex === 'number') {
        var feature = features && features[featureIndex];

        if (selectedFeatureId === (feature && feature.id)) {
          _this.setState({
            hoveredVertexIndex: vertexIndex
          });
        }
      } else if (type !== _constants.ELEMENT_TYPE.VERTEX) {
        _this.setState({
          hoveredVertexIndex: null
        });
      }

      if (type === _constants.ELEMENT_TYPE.FEATURE && typeof featureIndex === 'number') {
        var _feature = features && features[featureIndex];

        _this.setState({
          hoveredFeatureId: _feature && _feature.id
        });
      } else {
        _this.setState({
          hoveredFeatureId: null
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClickFeature", function (evt, attributes) {
      var featureIndex = attributes.featureIndex;
      var features = _this.state.features;
      var selectedFeature = features && typeof featureIndex === 'number' && features[featureIndex];

      if (selectedFeature) {
        _this.props.onSelect({
          selectedFeatureId: selectedFeature.id,
          sourceEvent: evt
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClickVertex", function (evt, attributes) {
      var mode = _this.props.mode;
      var operation = attributes.operation;

      if (operation === _constants.OPERATIONS.INTERSECT || operation === _constants.OPERATIONS.SET && mode === _constants.MODES.DRAW_RECTANGLE) {
        _this._closePath();

        _this._clearCache();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClickSegment", function (evt, attributes) {
      var feature = _this._getSelectedFeature();

      if (feature && (feature.renderType === _constants.RENDER_TYPE.POLYGON || feature.renderType === _constants.RENDER_TYPE.LINE_STRING) && attributes) {
        var vertexIndex = attributes.vertexIndex;
        var uncommittedLngLat = _this.state.uncommittedLngLat;
        var lngLat = uncommittedLngLat;

        if (!lngLat && typeof vertexIndex === 'number') {
          var _this$_getEventPositi3 = _this._getEventPosition(evt),
              x = _this$_getEventPositi3.x,
              y = _this$_getEventPositi3.y;

          lngLat = _this._unproject([x, y]);
          lngLat = _this._getClosestPositionOnSegment(vertexIndex, lngLat, feature);
        }

        if (lngLat) {
          var insertPosition = (vertexIndex + 1) % feature.points.length;
          feature.insertPoint(lngLat, insertPosition);

          _this._update(_this.state.features);
        }

        _this.setState({
          uncommittedLngLat: null,
          hoveredLngLat: null
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClick", function (evt) {
      var mode = _this.props.mode;
      var attributes = (0, _utils.parseElemDataAttributes)(evt.target);

      if (attributes && attributes.type === _constants.ELEMENT_TYPE.VERTEX) {
        _this._onClickVertex(evt, attributes);

        return;
      }

      if (mode === _constants.MODES.EDIT_VERTEX && attributes && attributes.type === _constants.ELEMENT_TYPE.SEGMENT) {
        _this._onClickSegment(evt, attributes);

        return;
      }

      if ((mode === _constants.MODES.SELECT_FEATURE || mode === _constants.MODES.EDIT_VERTEX) && attributes && attributes.type === _constants.ELEMENT_TYPE.FEATURE) {
        _this._onClickFeature(evt, attributes);

        return;
      }

      var selectedFeature = _this._getSelectedFeature();

      var _this$_getEventPositi4 = _this._getEventPosition(evt),
          x = _this$_getEventPositi4.x,
          y = _this$_getEventPositi4.y;

      switch (mode) {
        case _constants.MODES.EDIT_VERTEX:
          if (selectedFeature) {
            _this.props.onSelect({
              selectedFeatureId: null,
              sourceEvent: evt
            });
          }

          break;

        case _constants.MODES.DRAW_POINT:
          _this._addFeature(mode, {
            x: x,
            y: y
          });

          break;

        case _constants.MODES.DRAW_PATH:
        case _constants.MODES.DRAW_POLYGON:
          if (selectedFeature && selectedFeature.isClosed) {
            // clicked outside
            _this._clearCache();
          } else if (selectedFeature) {
            _this._addPoint(x, y, selectedFeature);
          } else {
            _this._addFeature(mode, {
              x: x,
              y: y
            });
          }

          break;

        case _constants.MODES.DRAW_RECTANGLE:
          if (selectedFeature && selectedFeature.isClosed) {
            // clicked outside
            _this._clearCache();

            _this.props.onSelect({
              selectedFeatureId: null,
              sourceEvent: evt
            });
          } else {
            _this._addFeature(mode, {
              x: x,
              y: y
            });
          }

          break;

        default:
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onPan", function (evt) {
      var attributes = (0, _utils.parseElemDataAttributes)(evt.target);
      var type = attributes && attributes.type;

      if (type === _constants.ELEMENT_TYPE.VERTEX || type === _constants.ELEMENT_TYPE.SEGMENT || _this.state.isDragging || _this.state.uncommittedLngLat !== null) {
        evt.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_project", function (pt) {
      return _this._context && _this._context.viewport && _this._context.viewport.project(pt);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_unproject", function (pt) {
      return _this._context && _this._context.viewport && _this._context.viewport.unproject(pt);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_matchesFeature", function (attributes, feature) {
      if (!attributes || attributes.type !== _constants.ELEMENT_TYPE.FEATURE || !feature) {
        return false;
      }

      var featureIndex = attributes.featureIndex;
      var features = _this.props.features;
      var elemFeature = features && features[featureIndex];
      return elemFeature && feature.id === elemFeature.id;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getClosestPositionOnSegment", function (vertexIndex, pointLngLat, feature) {
      var points = feature && feature.points;

      if (!points || !points.length) {
        return null;
      } // segmentId is start vertexIndex


      var startPos = points[vertexIndex];
      var endPos = points[(vertexIndex + 1) % points.length];
      return (0, _utils.findClosestPointOnLineSegment)(startPos, endPos, pointLngLat);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getEventPosition", function (evt) {
      var _evt$offsetCenter = evt.offsetCenter,
          x = _evt$offsetCenter.x,
          y = _evt$offsetCenter.y;
      return {
        x: Number(x),
        y: Number(y)
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getSelectedFeature", function () {
      var _this$state3 = _this.state,
          features = _this$state3.features,
          selectedFeatureId = _this$state3.selectedFeatureId;
      return features && features.find(function (f) {
        return f.id === selectedFeatureId;
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getEditHandleState", function (index, renderState) {
      var mode = _this.props.mode;
      var _this$state4 = _this.state,
          draggingVertexIndex = _this$state4.draggingVertexIndex,
          hoveredVertexIndex = _this$state4.hoveredVertexIndex;

      var selectedFeature = _this._getSelectedFeature();

      var isSelected = index === draggingVertexIndex || selectedFeature && selectedFeature.renderType === _constants.RENDER_TYPE.POINT;
      var isClosing = mode === _constants.MODES.DRAW_POLYGON && hoveredVertexIndex === 0 && index === -1;

      if (renderState) {
        return renderState;
      }

      if (isClosing) {
        return _constants.RENDER_STATE.CLOSING;
      }

      if (isSelected) {
        return _constants.RENDER_STATE.SELECTED;
      }

      switch (index) {
        case hoveredVertexIndex:
          return _constants.RENDER_STATE.HOVERED;

        case UNCOMMITTED_ID:
          return _constants.RENDER_STATE.UNCOMMITTED;

        default:
          return _constants.RENDER_STATE.INACTIVE;
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getFeatureRenderState", function (id, renderState) {
      var _this$state5 = _this.state,
          selectedFeatureId = _this$state5.selectedFeatureId,
          hoveredFeatureId = _this$state5.hoveredFeatureId;

      if (renderState) {
        return renderState;
      }

      switch (id) {
        case selectedFeatureId:
          return _constants.RENDER_STATE.SELECTED;

        case hoveredFeatureId:
          return _constants.RENDER_STATE.HOVERED;

        case UNCOMMITTED_ID:
          return _constants.RENDER_STATE.UNCOMMITTED;

        default:
          return _constants.RENDER_STATE.INACTIVE;
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderVertex", function (position, featureIndex, vertexIndex, operation, style, shape) {
      /* eslint-enable max-params */
      var p = _this._project(position);

      if (!p) {
        return null;
      }

      var clickRadius = _this.props.clickRadius;
      var elemKey = "".concat(_constants.ELEMENT_TYPE.VERTEX, ".").concat(featureIndex, ".").concat(vertexIndex, ".").concat(operation); // first <circle|rect> is to make path easily interacted with

      switch (shape) {
        case 'circle':
          return _react.default.createElement("g", {
            key: elemKey,
            transform: "translate(".concat(p[0], ", ").concat(p[1], ")")
          }, _react.default.createElement("circle", {
            "data-type": _constants.ELEMENT_TYPE.VERTEX,
            "data-feature-index": featureIndex,
            "data-vertex-index": vertexIndex,
            "data-operation": operation,
            key: "".concat(elemKey, ".hidden"),
            style: _objectSpread({}, style, _style.HIDDEN_CLICKABLE_STYLE, {
              stroke: 'none'
            }),
            cx: 0,
            cy: 0,
            r: clickRadius
          }), _react.default.createElement("circle", {
            "data-type": _constants.ELEMENT_TYPE.VERTEX,
            "data-feature-index": featureIndex,
            "data-vertex-index": vertexIndex,
            "data-operation": operation,
            key: elemKey,
            style: style,
            cx: 0,
            cy: 0
          }));

        case 'rect':
          return _react.default.createElement("g", {
            key: "vertex.".concat(vertexIndex),
            transform: "translate(".concat(p[0], ", ").concat(p[1], ")")
          }, _react.default.createElement("rect", {
            "data-type": _constants.ELEMENT_TYPE.VERTEX,
            "data-feature-index": featureIndex,
            "data-vertex-index": vertexIndex,
            "data-operation": operation,
            key: "".concat(elemKey, ".hidden"),
            style: _objectSpread({}, style, _style.HIDDEN_CLICKABLE_STYLE, {
              width: clickRadius,
              height: clickRadius
            })
          }), _react.default.createElement("rect", {
            "data-type": _constants.ELEMENT_TYPE.VERTEX,
            "data-feature-index": featureIndex,
            "data-vertex-index": vertexIndex,
            "data-operation": operation,
            key: elemKey,
            style: style
          }));

        default:
          return null;
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderSegment", function (featureIndex, startVertexId, startPos, endPos) {
      var style = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : {};
      var clickRadius = _this.props.clickRadius;

      var projected = _this._getProjectedData([startPos, endPos], _constants.RENDER_TYPE.LINE_STRING);

      var radius = style.radius,
          others = _objectWithoutProperties(style, ["radius"]);

      var elemKey = "".concat(_constants.ELEMENT_TYPE.SEGMENT, ".").concat(featureIndex, ".").concat(startVertexId);
      return _react.default.createElement("g", {
        key: elemKey
      }, _react.default.createElement("path", {
        "data-type": _constants.ELEMENT_TYPE.SEGMENT,
        "data-feature-index": featureIndex,
        "data-vertex-index": startVertexId,
        key: "".concat(elemKey, ".hidden"),
        style: _objectSpread({}, others, {
          strokeWidth: clickRadius || radius,
          opacity: 0
        }),
        d: projected
      }), _react.default.createElement("path", {
        "data-type": _constants.ELEMENT_TYPE.SEGMENT,
        "data-feature-index": featureIndex,
        "data-vertex-index": startVertexId,
        key: elemKey,
        style: others,
        d: projected
      }));
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCommittedStroke", function (featureIndex, feature, style) {
      var points = feature.points,
          isClosed = feature.isClosed,
          renderType = feature.renderType;

      if (!points || points.length < 2 || renderType === _constants.RENDER_TYPE.RECTANGLE && !isClosed) {
        return null;
      }

      var segments = [];

      for (var i = 0; i < points.length - 1; i++) {
        segments.push(_this._renderSegment(featureIndex, i, points[i], points[i + 1], style));
      }

      if (isClosed) {
        var lastIndex = points.length - 1;
        segments.push(_this._renderSegment(featureIndex, lastIndex, points[lastIndex], points[0], style));
      }

      return _react.default.createElement("g", {
        key: "committed group"
      }, segments);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderUncommittedStrokes", function (featureIndex, feature, style) {
      var points = feature.points,
          isClosed = feature.isClosed,
          renderType = feature.renderType;
      var mode = _this.props.mode;
      var uncommittedLngLat = _this.state.uncommittedLngLat;

      var isDrawing = _constants.DRAWING_MODES.find(function (m) {
        return m === mode;
      });

      if (!points || isClosed || !isDrawing) {
        return null;
      }

      var uncommittedSegments = [];

      if (renderType === _constants.RENDER_TYPE.RECTANGLE) {
        for (var i = 0; i < points.length - 1; i++) {
          uncommittedSegments.push(_this._renderSegment(featureIndex, i, points[i], points[i + 1], style));
        }

        if (points.length === 4) {
          uncommittedSegments.push(_this._renderSegment(featureIndex, 3, points[3], points[0], style));
        }
      }

      if (!uncommittedLngLat) {
        return uncommittedSegments.length ? uncommittedSegments : null;
      }
      /* eslint-disable no-inline-comments */


      uncommittedSegments.push(_this._renderSegment(featureIndex, points.length - 1, // id
      points.slice(-1)[0], // startPos
      uncommittedLngLat, // endPos
      style));
      /* eslint-enable no-inline-comments */

      return uncommittedSegments.filter(Boolean);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderClosingStroke", function (featureIndex, feature, style) {
      var points = feature.points,
          isClosed = feature.isClosed;
      var mode = _this.props.mode;
      var uncommittedLngLat = _this.state.uncommittedLngLat;

      if (uncommittedLngLat && mode === _constants.MODES.DRAW_POLYGON && points.length > 2 && !isClosed) {
        // from uncommitted position to the first point of the polygon
        return _this._renderSegment(featureIndex, 'uncommitted-close', uncommittedLngLat, points[0], style);
      }

      return null;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderFill", function (index, feature, style) {
      var mode = _this.props.mode;

      var isDrawing = _constants.DRAWING_MODES.find(function (m) {
        return m === mode;
      });

      var points = feature.points,
          renderType = feature.renderType,
          isClosed = feature.isClosed;

      if (renderType !== _constants.RENDER_TYPE.POLYGON && renderType !== _constants.RENDER_TYPE.RECTANGLE) {
        return null;
      }

      var uncommittedLngLat = _this.state.uncommittedLngLat;
      var fillPoints = points;

      if (uncommittedLngLat && isDrawing) {
        fillPoints = _toConsumableArray(points).concat([uncommittedLngLat]);
      }

      var fillPath = _this._getProjectedData(fillPoints, renderType, isClosed);

      return _react.default.createElement("path", {
        "data-type": _constants.ELEMENT_TYPE.FEATURE,
        "data-feature-index": index,
        key: "".concat(_constants.ELEMENT_TYPE.FEATURE, ".").concat(index, ".fill"),
        style: _objectSpread({}, style, {
          stroke: 'none'
        }),
        d: fillPath
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCurrentPath", function (feature, index) {
      var points = feature.points,
          renderType = feature.renderType;

      if (!points || !points.length || renderType === _constants.RENDER_TYPE.POINT) {
        return null;
      }

      var getFeatureStyle = _this.props.getFeatureStyle;
      var geoJson = feature.toFeature();
      var committedStyle = getFeatureStyle({
        feature: geoJson,
        state: _constants.RENDER_STATE.SELECTED
      });
      var uncommittedStyle = getFeatureStyle({
        feature: geoJson,
        state: _constants.RENDER_STATE.UNCOMMITTED
      });
      var closingStyle = getFeatureStyle({
        feature: geoJson,
        state: _constants.RENDER_STATE.CLOSING
      });

      var committedStroke = _this._renderCommittedStroke(index, feature, committedStyle);

      var uncommittedStrokes = _this._renderUncommittedStrokes(index, feature, uncommittedStyle) || [];

      var closingStroke = _this._renderClosingStroke(index, feature, closingStyle);

      var fill = _this._renderFill(index, feature, uncommittedStyle);

      return [fill, committedStroke].concat(_toConsumableArray(uncommittedStrokes), [closingStroke]).filter(Boolean);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCommittedVertices", function (featureIndex, feature, geoJson) {
      var _this$props2 = _this.props,
          mode = _this$props2.mode,
          getEditHandleStyle = _this$props2.getEditHandleStyle,
          getEditHandleShape = _this$props2.getEditHandleShape;
      var isClosed = feature.isClosed,
          points = feature.points;
      var committedVertices = [];

      for (var i = 0; i < points.length; i++) {
        var p = points[i];
        var operation = _constants.OPERATIONS.SET;
        var style = getEditHandleStyle({
          feature: geoJson,
          index: i,
          state: _this._getEditHandleState(i)
        });
        var shape = typeof getEditHandleShape === 'function' ? getEditHandleShape({
          feature: geoJson,
          index: i,
          state: _this._getEditHandleState(i)
        }) : getEditHandleShape;

        if (isClosed) {
          committedVertices.push(_this._renderVertex(p, featureIndex, i, operation, style, shape));
        } else {
          if (mode === _constants.MODES.DRAW_POLYGON && i === 0 && points.length > 2) {
            operation = _constants.OPERATIONS.INTERSECT;
          }

          committedVertices.push(_this._renderVertex(p, featureIndex, i, operation, style, shape));
        }
      }

      return committedVertices;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderUncommittedVertex", function (featureIndex, feature, geoJson) {
      var _this$props3 = _this.props,
          getEditHandleStyle = _this$props3.getEditHandleStyle,
          getEditHandleShape = _this$props3.getEditHandleShape;
      var _this$state6 = _this.state,
          selectedFeatureId = _this$state6.selectedFeatureId,
          uncommittedLngLat = _this$state6.uncommittedLngLat;
      var id = feature.id;
      var uncommittedVertex = null;

      if (selectedFeatureId === id && uncommittedLngLat) {
        var style = getEditHandleStyle({
          feature: geoJson,
          index: 'uncommitted',
          state: _this._getEditHandleState(-1, _constants.RENDER_STATE.UNCOMMITTED)
        });
        var shape = typeof getEditHandleShape === 'function' ? getEditHandleShape({
          feature: geoJson,
          index: null,
          state: _this._getEditHandleState(-1)
        }) : getEditHandleShape;
        uncommittedVertex = _this._renderVertex(uncommittedLngLat, featureIndex, 'uncommitted', _constants.OPERATIONS.INSERT, _objectSpread({}, style, {
          pointerEvents: 'none'
        }), shape);
      }

      return uncommittedVertex;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCurrentVertices", function (feature, featureIndex) {
      var points = feature.points;

      if (!points || !points.length) {
        return null;
      }

      var geoJson = feature.toFeature();

      var committedVertices = _this._renderCommittedVertices(featureIndex, feature, geoJson);

      var uncommittedVertex = _this._renderUncommittedVertex(featureIndex, feature, geoJson);

      return _react.default.createElement("g", {
        key: "edit-handles"
      }, committedVertices, uncommittedVertex);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCurrent", function () {
      var features = _this.state.features;

      var feature = _this._getSelectedFeature();

      if (!features || !feature || !feature.points) {
        return null;
      }

      var mode = _this.props.mode;
      var index = features.findIndex(function (f) {
        return f.id === feature.id;
      });
      return _react.default.createElement("g", {
        key: "feature current",
        style: mode === _constants.MODES.READ_ONLY || mode === _constants.MODES.SELECT_FEATURE ? _constants.STATIC_STYLE : null
      }, _this._renderCurrentPath(feature, index), _this._renderCurrentVertices(feature, index));
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderFeature", function (feature, index) {
      if (feature === _this._getSelectedFeature()) {
        return null;
      }

      var clickRadius = _this.props.clickRadius;
      var id = feature.id,
          points = feature.points,
          renderType = feature.renderType,
          isClosed = feature.isClosed;

      if (!points || !points.length) {
        return null;
      }

      var projected = _this._getProjectedData(points, renderType, isClosed);

      if (!projected) {
        return null;
      }

      var geoJson = feature.toFeature();

      var renderState = _this._getFeatureRenderState(id);

      var _this$props4 = _this.props,
          getFeatureStyle = _this$props4.getFeatureStyle,
          getFeatureShape = _this$props4.getFeatureShape;
      var style = getFeatureStyle({
        feature: geoJson,
        state: renderState
      });
      var shape = typeof getFeatureShape === 'function' ? getFeatureShape({
        feature: geoJson,
        state: renderState
      }) : getFeatureShape;
      var elemKey = "".concat(_constants.ELEMENT_TYPE.FEATURE, ".").concat(index);

      switch (renderType) {
        case _constants.RENDER_TYPE.POINT:
          if (shape === 'rect') {
            return _react.default.createElement("g", {
              key: elemKey,
              transform: "translate(".concat(projected[0][0], ", ").concat(projected[0][1], ")")
            }, _react.default.createElement("rect", {
              "data-type": _constants.ELEMENT_TYPE.FEATURE,
              "data-feature-index": index,
              key: "".concat(elemKey, ".hidden"),
              style: _objectSpread({}, style, _style.HIDDEN_CLICKABLE_STYLE, {
                width: clickRadius,
                height: clickRadius
              })
            }), _react.default.createElement("rect", {
              "data-type": _constants.ELEMENT_TYPE.FEATURE,
              "data-feature-index": index,
              key: elemKey,
              style: style
            }));
          }

          return _react.default.createElement("g", {
            key: elemKey,
            transform: "translate(".concat(projected[0][0], ", ").concat(projected[0][1], ")")
          }, _react.default.createElement("circle", {
            "data-type": _constants.ELEMENT_TYPE.FEATURE,
            "data-feature-index": index,
            key: "".concat(elemKey, ".hidden"),
            style: _objectSpread({}, style, {
              opacity: 0
            }),
            cx: 0,
            cy: 0,
            r: clickRadius
          }), _react.default.createElement("circle", {
            "data-type": _constants.ELEMENT_TYPE.FEATURE,
            "data-feature-index": index,
            key: "feature.".concat(index),
            style: style,
            cx: 0,
            cy: 0
          }));
        // first <path> is to make path easily interacted with

        case _constants.RENDER_TYPE.LINE_STRING:
          return _react.default.createElement("g", {
            key: elemKey
          }, _react.default.createElement("path", {
            "data-type": _constants.ELEMENT_TYPE.FEATURE,
            "data-feature-index": index,
            key: "".concat(elemKey, ".hidden"),
            style: _objectSpread({}, style, {
              strokeWidth: clickRadius,
              opacity: 0
            }),
            d: projected
          }), _react.default.createElement("path", {
            "data-type": _constants.ELEMENT_TYPE.FEATURE,
            "data-feature-index": index,
            key: elemKey,
            style: style,
            d: projected
          }));

        case 'Polygon':
        case 'Rectangle':
          return _react.default.createElement("path", {
            "data-type": _constants.ELEMENT_TYPE.FEATURE,
            "data-feature-index": index,
            key: elemKey,
            style: style,
            d: projected
          });

        default:
          return null;
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderFeatures", function () {
      var features = _this.state.features;
      return features && features.map(_this._renderFeature);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderCanvas", function () {
      var _this$state7 = _this.state,
          selectedFeatureId = _this$state7.selectedFeatureId,
          features = _this$state7.features;
      return _react.default.createElement("svg", {
        key: "draw-canvas",
        width: "100%",
        height: "100%"
      }, features && features.length > 0 && _react.default.createElement("g", {
        key: "feature-group"
      }, _this._renderFeatures()), selectedFeatureId && _this._renderCurrent());
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderEditor", function () {
      var viewport = _this._context && _this._context.viewport || {};
      var style = _this.props.style;
      var width = viewport.width,
          height = viewport.height;
      return _react.default.createElement("div", {
        id: "editor",
        style: _objectSpread({
          width: width,
          height: height
        }, style),
        ref: function ref(_) {
          _this._containerRef = _;
        }
      }, _this._renderCanvas());
    });

    _this.state = {
      features: props.features ? props.features.map(function (f) {
        return _feature2.default.fromFeature(f);
      }).filter(Boolean) : null,
      selectedFeatureId: -1,
      hoveredFeatureId: null,
      hoveredLngLat: null,
      hoveredVertexIndex: -1,
      // intermediate mouse position when drawing
      uncommittedLngLat: null,
      draggingVertexIndex: -1,
      startDragPos: null,
      isDragging: false,
      didDrag: false
    };
    _this._containerRef = null;
    _this._events = {};
    _this._context = null;
    return _this;
  }

  _createClass(Editor, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.mode && this.props.mode !== _constants.MODES.READ_ONLY) {
        this._setupEvents();
      }
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.mode !== nextProps.mode) {
        if (!nextProps.mode || nextProps.mode === _constants.MODES.READ_ONLY) {
          this._removeEvents();
        }

        if (!this.props.mode || this.props.mode === _constants.MODES.READ_ONLY) {
          this._setupEvents();
        }
      }

      if (this.props.mode !== nextProps.mode || this.props.features !== nextProps.features) {
        this.setState({
          features: nextProps.features && nextProps.features.map(function (f) {
            return _feature2.default.fromFeature(f);
          }).filter(Boolean)
        });
      }

      if (this.props.mode !== nextProps.mode || this.props.selectedFeatureId !== nextProps.selectedFeatureId) {
        this._clearCache();

        this.setState({
          selectedFeatureId: nextProps.selectedFeatureId
        });
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._removeEvents();
    }
  }, {
    key: "_setupEvents",

    /* EVENTS */
    value: function _setupEvents() {
      var _this2 = this;

      var ref = this._containerRef;

      if (!ref || !this._context || !this._context.eventManager) {
        return;
      }

      this._events = {
        anyclick: function anyclick(evt) {
          return _this2._onEvent(_this2._onClick, evt, true);
        },
        click: function click(evt) {
          return evt.stopImmediatePropagation();
        },
        pointermove: function pointermove(evt) {
          return _this2._onEvent(_this2._onMouseMove, evt, true);
        },
        pointerdown: function pointerdown(evt) {
          return _this2._onEvent(_this2._onMouseDown, evt, true);
        },
        pointerup: function pointerup(evt) {
          return _this2._onEvent(_this2._onMouseUp, evt, true);
        },
        panmove: function panmove(evt) {
          return _this2._onEvent(_this2._onPan, evt, false);
        },
        panstart: function panstart(evt) {
          return _this2._onEvent(_this2._onPan, evt, false);
        },
        panend: function panend(evt) {
          return _this2._onEvent(_this2._onPan, evt, false);
        }
      };

      this._context.eventManager.on(this._events, ref);
    }
  }, {
    key: "_removeEvents",
    value: function _removeEvents() {
      if (!this._context || !this._context.eventManager || !this._events) {
        return;
      }

      this._context.eventManager.off(this._events);

      this._events = null;
    }
  }, {
    key: "_getProjectedData",
    value: function _getProjectedData(points, renderType) {
      var _this3 = this;

      var isClosed = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

      if (points.length === 0) {
        return '';
      }

      var projected = points.map(function (p) {
        return _this3._project(p);
      });

      switch (renderType) {
        case _constants.RENDER_TYPE.POINT:
          return projected;

        case _constants.RENDER_TYPE.LINE_STRING:
        case _constants.RENDER_TYPE.POLYGON:
        case _constants.RENDER_TYPE.RECTANGLE:
          var pathString = projected.map(function (p) {
            return "".concat(p[0], ",").concat(p[1]);
          }).join('L');
          return "M ".concat(pathString, " ").concat(isClosed ? 'z' : '');

        default:
          return null;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      return _react.default.createElement(_reactMapGl._MapContext.Consumer, null, function (context) {
        _this4._context = context;
        var viewport = context && context.viewport;

        if (!viewport || viewport.height <= 0 || viewport.width <= 0) {
          return null;
        }

        return _this4._renderEditor();
      });
    }
  }]);

  return Editor;
}(_react.PureComponent);

exports.default = Editor;

_defineProperty(Editor, "defaultProps", defaultProps);
