// @loaders.gl/wms/dist/esm/lib/data-sources/adhoc-image-service.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

import { ImageLoader } from '@loaders.gl/images';
import { ImageSource } from './image-source';
export class AdHocImageService extends ImageSource {
  constructor(props) {
    super();
    _defineProperty(this, "templateUrl", void 0);
    this.templateUrl = props.templateUrl;
  }

  async getMetadata() {
    throw new Error('ImageSource.getMetadata not implemented');
  }
  async getImage(parameters) {
    const granularParameters = this.getGranularParameters(parameters);
    const url = this.getURLFromTemplate(granularParameters);
    const response = await fetch(url);
    const arrayBuffer = await response.arrayBuffer();
    return await ImageLoader.parse(arrayBuffer);
  }

  getGranularParameters(parameters) {
    const [east, north, west, south] = parameters.bbox;
    return {
      ...parameters,
      east,
      north,
      south,
      west
    };
  }
  getURLFromTemplate(parameters) {
    let url = this.templateUrl;
    for (const [key, value] of Object.entries(parameters)) {
      url = url.replace("${".concat(key, "}"), String(value));
    }
    return url;
  }
}
_defineProperty(AdHocImageService, "type", 'template');
_defineProperty(AdHocImageService, "testURL", url => url.toLowerCase().includes('{'));
