// @kepler.gl/components/dist/side-panel/common/dataset-tag.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DatasetTagFactory;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _localization = require("@kepler.gl/localization");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = require("../..");

var _templateObject, _templateObject2;

var DatasetTagWrapper = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  color: ", ";\n  font-size: 11px;\n  letter-spacing: 0.2px;\n  overflow: auto;\n\n  .dataset-color {\n    flex-shrink: 0;\n    margin-top: 3px;\n  }\n\n  .dataset-name {\n    white-space: nowrap;\n    overflow: hidden;\n    text-overflow: ellipsis;\n  }\n"])), function (props) {
  return props.theme.textColor;
});

var DatasetColorPicker = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n"])));

var UpdateTableColor = function UpdateTableColor(_ref) {
  var children = _ref.children,
      id = _ref.id;
  return /*#__PURE__*/_react["default"].createElement(DatasetColorPicker, {
    className: "dataset-action update-color",
    "data-tip": true,
    "data-for": "update-color-".concat(id)
  }, children);
};

function DatasetTagFactory() {
  var DatasetTag = function DatasetTag(_ref2) {
    var onClick = _ref2.onClick,
        onClickSquare = _ref2.onClickSquare,
        dataset = _ref2.dataset,
        updateTableColor = _ref2.updateTableColor,
        id = _ref2.id;
    return /*#__PURE__*/_react["default"].createElement(DatasetTagWrapper, {
      className: "source-data-tag"
    }, /*#__PURE__*/_react["default"].createElement(UpdateTableColor, {
      id: dataset.id
    }, /*#__PURE__*/_react["default"].createElement(_.DatasetSquare, {
      className: "dataset-color",
      backgroundColor: dataset.color,
      onClick: onClickSquare,
      "data-tip": true,
      "data-for": "update-color-".concat(id)
    }), updateTableColor ? /*#__PURE__*/_react["default"].createElement(_.Tooltip, {
      id: "update-color-".concat(id),
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: 'Update color'
    }))) : null), /*#__PURE__*/_react["default"].createElement("div", {
      className: "dataset-name",
      title: dataset.label,
      onClick: onClick
    }, dataset.label));
  };

  return DatasetTag;
}
