// @kepler.gl/components/dist/side-panel/common/source-data-selector.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = SourceDataSelectorFactory;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = require("../../common/styled-components");

var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));

var _datasetTag = _interopRequireDefault(require("./dataset-tag"));

var _localization = require("@kepler.gl/localization");var defaultPlaceHolder = 'Select A Dataset';
SourceDataSelectorFactory.deps = [_datasetTag["default"]];

function SourceDataSelectorFactory(DatasetTag) {
  var DatasetItem = function DatasetItem(_ref) {
    var value = _ref.value;
    return /*#__PURE__*/_react["default"].createElement(DatasetTag, {
      dataset: value
    });
  };

  var SourceDataSelector = function SourceDataSelector(_ref2) {
    var dataId = _ref2.dataId,
        datasets = _ref2.datasets,
        disabled = _ref2.disabled,
        onSelect = _ref2.onSelect,
        defaultValue = _ref2.defaultValue,
        inputTheme = _ref2.inputTheme;
    var dsOptions = (0, _react.useMemo)(function () {
      return Object.values(datasets).map(function (ds) {
        return {
          label: ds.label,
          value: ds.id,
          color: ds.color
        };
      });
    }, [datasets]);
    var selectedItems = (0, _react.useMemo)(function () {
      return dataId ? (Array.isArray(dataId) && dataId || [dataId]).map(function (id) {
        return datasets[id];
      }) : [];
    }, [dataId, datasets]);
    return /*#__PURE__*/_react["default"].createElement(_styledComponents.SidePanelSection, {
      className: "data-source-selector"
    }, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabel, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: 'misc.dataSource'
    })), /*#__PURE__*/_react["default"].createElement(_itemSelector["default"], {
      inputTheme: inputTheme,
      selectedItems: selectedItems,
      options: dsOptions,
      getOptionValue: 'value',
      filterOption: 'label',
      multiSelect: false,
      onChange: onSelect,
      placeholder: defaultValue,
      disabled: Boolean(disabled),
      displayOption: 'label',
      DropDownLineItemRenderComponent: DatasetItem
    }));
  };

  SourceDataSelector.defaultProps = {
    defaultValue: defaultPlaceHolder
  };
  return SourceDataSelector;
}
