// @kepler.gl/schemas/dist/index.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  KeplerGlSchema: true,
  reducerSchema: true,
  KeplerGLSchemaClass: true,
  CURRENT_VERSION: true,
  VERSIONS: true,
  visStateSchemaV1: true,
  FilterSchemaV0: true,
  LayerSchemaV0: true,
  InteractionSchemaV1: true,
  DimensionFieldSchema: true,
  SplitMapsSchema: true,
  filterPropsV1: true,
  visStateSchema: true,
  datasetSchema: true,
  DatasetSchema: true,
  fieldPropertiesV1: true,
  datasetPropertiesV1: true,
  mapStyleSchema: true,
  mapStateSchema: true,
  Schema: true
};
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _schemaManager["default"];
  }
});
Object.defineProperty(exports, "KeplerGlSchema", {
  enumerable: true,
  get: function get() {
    return _schemaManager["default"];
  }
});
Object.defineProperty(exports, "reducerSchema", {
  enumerable: true,
  get: function get() {
    return _schemaManager.reducerSchema;
  }
});
Object.defineProperty(exports, "KeplerGLSchemaClass", {
  enumerable: true,
  get: function get() {
    return _schemaManager.KeplerGLSchema;
  }
});
Object.defineProperty(exports, "CURRENT_VERSION", {
  enumerable: true,
  get: function get() {
    return _versions.CURRENT_VERSION;
  }
});
Object.defineProperty(exports, "VERSIONS", {
  enumerable: true,
  get: function get() {
    return _versions.VERSIONS;
  }
});
Object.defineProperty(exports, "visStateSchemaV1", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.visStateSchemaV1;
  }
});
Object.defineProperty(exports, "FilterSchemaV0", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.FilterSchemaV0;
  }
});
Object.defineProperty(exports, "LayerSchemaV0", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.LayerSchemaV0;
  }
});
Object.defineProperty(exports, "InteractionSchemaV1", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.InteractionSchemaV1;
  }
});
Object.defineProperty(exports, "DimensionFieldSchema", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.DimensionFieldSchema;
  }
});
Object.defineProperty(exports, "SplitMapsSchema", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.SplitMapsSchema;
  }
});
Object.defineProperty(exports, "filterPropsV1", {
  enumerable: true,
  get: function get() {
    return _visStateSchema.filterPropsV1;
  }
});
Object.defineProperty(exports, "visStateSchema", {
  enumerable: true,
  get: function get() {
    return _visStateSchema["default"];
  }
});
Object.defineProperty(exports, "datasetSchema", {
  enumerable: true,
  get: function get() {
    return _datasetSchema["default"];
  }
});
Object.defineProperty(exports, "DatasetSchema", {
  enumerable: true,
  get: function get() {
    return _datasetSchema.DatasetSchema;
  }
});
Object.defineProperty(exports, "fieldPropertiesV1", {
  enumerable: true,
  get: function get() {
    return _datasetSchema.fieldPropertiesV1;
  }
});
Object.defineProperty(exports, "datasetPropertiesV1", {
  enumerable: true,
  get: function get() {
    return _datasetSchema.propertiesV1;
  }
});
Object.defineProperty(exports, "mapStyleSchema", {
  enumerable: true,
  get: function get() {
    return _mapStyleSchema["default"];
  }
});
Object.defineProperty(exports, "mapStateSchema", {
  enumerable: true,
  get: function get() {
    return _mapStateSchema["default"];
  }
});
Object.defineProperty(exports, "Schema", {
  enumerable: true,
  get: function get() {
    return _schema["default"];
  }
});

var _schemaManager = _interopRequireWildcard(require("./schema-manager"));

var _versions = require("./versions");

var _visStateSchema = _interopRequireWildcard(require("./vis-state-schema"));

Object.keys(_visStateSchema).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _visStateSchema[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _visStateSchema[key];
    }
  });
});

var _datasetSchema = _interopRequireWildcard(require("./dataset-schema"));

var _mapStyleSchema = _interopRequireDefault(require("./map-style-schema"));

var _mapStateSchema = _interopRequireDefault(require("./map-state-schema"));

var _schema = _interopRequireDefault(require("./schema"));
