// @loaders.gl/json/dist/esm/geojson-loader.js
import { geojsonToBinary } from '@loaders.gl/gis';
import parseJSONSync from './lib/parse-json';
import parseJSONInBatches from './lib/parse-json-in-batches';

const VERSION = typeof "3.3.3" !== 'undefined' ? "3.3.3" : 'latest';
const DEFAULT_GEOJSON_LOADER_OPTIONS = {
  geojson: {
    shape: 'object-row-table'
  },
  json: {
    jsonpaths: ['$', '$.features']
  },
  gis: {
    format: 'geojson'
  }
};

export const GeoJSONWorkerLoader = {
  name: 'GeoJSON',
  id: 'geojson',
  module: 'geojson',
  version: VERSION,
  worker: true,
  extensions: ['geojson'],
  mimeTypes: ['application/geo+json'],
  category: 'geometry',
  text: true,
  options: DEFAULT_GEOJSON_LOADER_OPTIONS
};
export const GeoJSONLoader = {
  ...GeoJSONWorkerLoader,
  parse,
  parseTextSync,
  parseInBatches
};
async function parse(arrayBuffer, options) {
  return parseTextSync(new TextDecoder().decode(arrayBuffer), options);
}
function parseTextSync(text, options) {
  options = {
    ...DEFAULT_GEOJSON_LOADER_OPTIONS,
    ...options
  };
  options.json = {
    ...DEFAULT_GEOJSON_LOADER_OPTIONS.geojson,
    ...options.geojson
  };
  options.gis = options.gis || {};
  const json = parseJSONSync(text, options);
  switch (options.gis.format) {
    case 'binary':
      return geojsonToBinary(json);
    default:
      return json;
  }
}
function parseInBatches(asyncIterator, options) {
  options = {
    ...DEFAULT_GEOJSON_LOADER_OPTIONS,
    ...options
  };
  options.json = {
    ...DEFAULT_GEOJSON_LOADER_OPTIONS.geojson,
    ...options.geojson
  };
  const geojsonIterator = parseJSONInBatches(asyncIterator, options);
  switch (options.gis.format) {
    case 'binary':
      return makeBinaryGeometryIterator(geojsonIterator);
    default:
      return geojsonIterator;
  }
}
async function* makeBinaryGeometryIterator(geojsonIterator) {
  for await (const batch of geojsonIterator) {
    batch.data = geojsonToBinary(batch.data);
    yield batch;
  }
}
