// @salte-auth/popup/dist/popup.js

/**
 * @salte-auth/popup JavaScript Library v1.0.0-rc.1
 *
 * @license MIT (https://github.com/salte-auth/popup/blob/master/LICENSE)
 *
 * Made with ♥ by Nick Woodward <nick@salte.io>, Dave Woodward <dave@salte.io>
 */

(function (factory) {
	typeof define === 'function' && define.amd ? define(factory) :
	factory();
}(function () { 'use strict';

	var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

	function unwrapExports (x) {
		return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
	}

	function createCommonjsModule(fn, module) {
		return module = { exports: {} }, fn(module, module.exports), module.exports;
	}

	var salteAuth_min = createCommonjsModule(function (module, exports) {
	/**
	 * @salte-auth/salte-auth JavaScript Library v3.0.0-rc.5
	 *
	 * @license MIT (https://github.com/salte-auth/salte-auth/blob/master/LICENSE)
	 *
	 * Made with ♥ by Nick Woodward <nick@salte.io>, Dave Woodward <dave@salte.io>
	 */
	!function(e,t){t(exports);}(commonjsGlobal,function(e){function t(e,t){return e(t={exports:{}},t.exports),t.exports}var n=t(function(e){var t=function(e){var t,n=Object.prototype,r=n.hasOwnProperty,o="function"==typeof Symbol?Symbol:{},i=o.iterator||"@@iterator",a=o.asyncIterator||"@@asyncIterator",c=o.toStringTag||"@@toStringTag";function s(e,t,n,r){var o=t&&t.prototype instanceof v?t:v,i=Object.create(o.prototype),a=new L(r||[]);return i._invoke=function(e,t,n){var r=l;return function(o,i){if(r===h)throw new Error("Generator is already running");if(r===p){if("throw"===o)throw i;return T()}for(n.method=o,n.arg=i;;){var a=n.delegate;if(a){var c=E(a,n);if(c){if(c===d)continue;return c}}if("next"===n.method)n.sent=n._sent=n.arg;else if("throw"===n.method){if(r===l)throw r=p,n.arg;n.dispatchException(n.arg);}else"return"===n.method&&n.abrupt("return",n.arg);r=h;var s=u(e,t,n);if("normal"===s.type){if(r=n.done?p:f,s.arg===d)continue;return {value:s.arg,done:n.done}}"throw"===s.type&&(r=p,n.method="throw",n.arg=s.arg);}}}(e,n,a),i}function u(e,t,n){try{return {type:"normal",arg:e.call(t,n)}}catch(e){return {type:"throw",arg:e}}}e.wrap=s;var l="suspendedStart",f="suspendedYield",h="executing",p="completed",d={};function v(){}function y(){}function g(){}var m={};m[i]=function(){return this};var w=Object.getPrototypeOf,k=w&&w(w(A([])));k&&k!==n&&r.call(k,i)&&(m=k);var b=g.prototype=v.prototype=Object.create(m);function x(e){["next","throw","return"].forEach(function(t){e[t]=function(e){return this._invoke(t,e)};});}function _(e){var t;this._invoke=function(n,o){function i(){return new Promise(function(t,i){!function t(n,o,i,a){var c=u(e[n],e,o);if("throw"!==c.type){var s=c.arg,l=s.value;return l&&"object"==typeof l&&r.call(l,"__await")?Promise.resolve(l.__await).then(function(e){t("next",e,i,a);},function(e){t("throw",e,i,a);}):Promise.resolve(l).then(function(e){s.value=e,i(s);},function(e){return t("throw",e,i,a)})}a(c.arg);}(n,o,t,i);})}return t=t?t.then(i,i):i()};}function E(e,n){var r=e.iterator[n.method];if(r===t){if(n.delegate=null,"throw"===n.method){if(e.iterator.return&&(n.method="return",n.arg=t,E(e,n),"throw"===n.method))return d;n.method="throw",n.arg=new TypeError("The iterator does not provide a 'throw' method");}return d}var o=u(r,e.iterator,n.arg);if("throw"===o.type)return n.method="throw",n.arg=o.arg,n.delegate=null,d;var i=o.arg;return i?i.done?(n[e.resultName]=i.value,n.next=e.nextLoc,"return"!==n.method&&(n.method="next",n.arg=t),n.delegate=null,d):i:(n.method="throw",n.arg=new TypeError("iterator result is not an object"),n.delegate=null,d)}function O(e){var t={tryLoc:e[0]};1 in e&&(t.catchLoc=e[1]),2 in e&&(t.finallyLoc=e[2],t.afterLoc=e[3]),this.tryEntries.push(t);}function S(e){var t=e.completion||{};t.type="normal",delete t.arg,e.completion=t;}function L(e){this.tryEntries=[{tryLoc:"root"}],e.forEach(O,this),this.reset(!0);}function A(e){if(e){var n=e[i];if(n)return n.call(e);if("function"==typeof e.next)return e;if(!isNaN(e.length)){var o=-1,a=function n(){for(;++o<e.length;)if(r.call(e,o))return n.value=e[o],n.done=!1,n;return n.value=t,n.done=!0,n};return a.next=a}}return {next:T}}function T(){return {value:t,done:!0}}return y.prototype=b.constructor=g,g.constructor=y,g[c]=y.displayName="GeneratorFunction",e.isGeneratorFunction=function(e){var t="function"==typeof e&&e.constructor;return !!t&&(t===y||"GeneratorFunction"===(t.displayName||t.name))},e.mark=function(e){return Object.setPrototypeOf?Object.setPrototypeOf(e,g):(e.__proto__=g,c in e||(e[c]="GeneratorFunction")),e.prototype=Object.create(b),e},e.awrap=function(e){return {__await:e}},x(_.prototype),_.prototype[a]=function(){return this},e.AsyncIterator=_,e.async=function(t,n,r,o){var i=new _(s(t,n,r,o));return e.isGeneratorFunction(n)?i:i.next().then(function(e){return e.done?e.value:i.next()})},x(b),b[c]="Generator",b[i]=function(){return this},b.toString=function(){return "[object Generator]"},e.keys=function(e){var t=[];for(var n in e)t.push(n);return t.reverse(),function n(){for(;t.length;){var r=t.pop();if(r in e)return n.value=r,n.done=!1,n}return n.done=!0,n}},e.values=A,L.prototype={constructor:L,reset:function(e){if(this.prev=0,this.next=0,this.sent=this._sent=t,this.done=!1,this.delegate=null,this.method="next",this.arg=t,this.tryEntries.forEach(S),!e)for(var n in this)"t"===n.charAt(0)&&r.call(this,n)&&!isNaN(+n.slice(1))&&(this[n]=t);},stop:function(){this.done=!0;var e=this.tryEntries[0].completion;if("throw"===e.type)throw e.arg;return this.rval},dispatchException:function(e){if(this.done)throw e;var n=this;function o(r,o){return c.type="throw",c.arg=e,n.next=r,o&&(n.method="next",n.arg=t),!!o}for(var i=this.tryEntries.length-1;i>=0;--i){var a=this.tryEntries[i],c=a.completion;if("root"===a.tryLoc)return o("end");if(a.tryLoc<=this.prev){var s=r.call(a,"catchLoc"),u=r.call(a,"finallyLoc");if(s&&u){if(this.prev<a.catchLoc)return o(a.catchLoc,!0);if(this.prev<a.finallyLoc)return o(a.finallyLoc)}else if(s){if(this.prev<a.catchLoc)return o(a.catchLoc,!0)}else{if(!u)throw new Error("try statement without catch or finally");if(this.prev<a.finallyLoc)return o(a.finallyLoc)}}}},abrupt:function(e,t){for(var n=this.tryEntries.length-1;n>=0;--n){var o=this.tryEntries[n];if(o.tryLoc<=this.prev&&r.call(o,"finallyLoc")&&this.prev<o.finallyLoc){var i=o;break}}i&&("break"===e||"continue"===e)&&i.tryLoc<=t&&t<=i.finallyLoc&&(i=null);var a=i?i.completion:{};return a.type=e,a.arg=t,i?(this.method="next",this.next=i.finallyLoc,d):this.complete(a)},complete:function(e,t){if("throw"===e.type)throw e.arg;return "break"===e.type||"continue"===e.type?this.next=e.arg:"return"===e.type?(this.rval=this.arg=e.arg,this.method="return",this.next="end"):"normal"===e.type&&t&&(this.next=t),d},finish:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var n=this.tryEntries[t];if(n.finallyLoc===e)return this.complete(n.completion,n.afterLoc),S(n),d}},catch:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var n=this.tryEntries[t];if(n.tryLoc===e){var r=n.completion;if("throw"===r.type){var o=r.arg;S(n);}return o}}throw new Error("illegal catch attempt")},delegateYield:function(e,n,r){return this.delegate={iterator:A(e),resultName:n,nextLoc:r},"next"===this.method&&(this.arg=t),d}},e}(e.exports);try{regeneratorRuntime=t;}catch(e){Function("r","regeneratorRuntime = r")(t);}});function r(e,t,n,r,o,i,a){try{var c=e[i](a),s=c.value;}catch(e){return void n(e)}c.done?t(s):Promise.resolve(s).then(r,o);}var o=function(e){return function(){var t=this,n=arguments;return new Promise(function(o,i){var a=e.apply(t,n);function c(e){r(a,o,i,c,s,"next",e);}function s(e){r(a,o,i,c,s,"throw",e);}c(void 0);})}};var i=function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")};function a(e,t){for(var n=0;n<t.length;n++){var r=t[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(e,r.key,r);}}var c=function(e,t,n){return t&&a(e.prototype,t),n&&a(e,n),e},s=t(function(e){function t(e){return (t="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}function n(r){return "function"==typeof Symbol&&"symbol"===t(Symbol.iterator)?e.exports=n=function(e){return t(e)}:e.exports=n=function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":t(e)},n(r)}e.exports=n;});var u=function(e){if(void 0===e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return e};var l=function(e,t){return !t||"object"!==s(t)&&"function"!=typeof t?u(e):t},f=t(function(e){function t(n){return e.exports=t=Object.setPrototypeOf?Object.getPrototypeOf:function(e){return e.__proto__||Object.getPrototypeOf(e)},t(n)}e.exports=t;}),h=t(function(e){function t(n,r){return e.exports=t=Object.setPrototypeOf||function(e,t){return e.__proto__=t,e},t(n,r)}e.exports=t;});var p=function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function");e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,writable:!0,configurable:!0}}),t&&h(e,t);};var d=function(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e};var v=function(e){return -1!==Function.toString.call(e).indexOf("[native code]")},y=t(function(e){function t(n,r,o){return !function(){if("undefined"==typeof Reflect||!Reflect.construct)return !1;if(Reflect.construct.sham)return !1;if("function"==typeof Proxy)return !0;try{return Date.prototype.toString.call(Reflect.construct(Date,[],function(){})),!0}catch(e){return !1}}()?e.exports=t=function(e,t,n){var r=[null];r.push.apply(r,t);var o=new(Function.bind.apply(e,r));return n&&h(o,n.prototype),o}:e.exports=t=Reflect.construct,t.apply(null,arguments)}e.exports=t;}),g=function(e){function t(e){var n,r=e.message,o=e.code;return i(this,t),n=l(this,f(t).call(this,r)),d(u(n),"code",void 0),n.code=o,n}return p(t,e),t}(t(function(e){function t(n){var r="function"==typeof Map?new Map:void 0;return e.exports=t=function(e){if(null===e||!v(e))return e;if("function"!=typeof e)throw new TypeError("Super expression must either be null or a function");if(void 0!==r){if(r.has(e))return r.get(e);r.set(e,t);}function t(){return y(e,arguments,f(this).constructor)}return t.prototype=Object.create(e.prototype,{constructor:{value:t,enumerable:!1,writable:!0,configurable:!0}}),h(t,e)},t(n)}e.exports=t;})(Error)),m=function(){function e(t){i(this,e),this.config=t||{};}return c(e,[{key:"required",value:function(){for(var e=this,t=arguments.length,n=new Array(t),r=0;r<t;r++)n[r]=arguments[r];var o=n.filter(function(t){return void 0===e.config[t]});if(o.length>0)throw new g({code:"missing_required_properties",message:"Missing the following required fields. (".concat(o.join(", "),")")})}}]),e}(),w=function(){function e(){i(this,e);}return c(e,null,[{key:"setup",value:function(t){this.hasSetup&&!t||(this.hasSetup=!0,this.interceptors=[],this.real||(this.real=window.fetch),window.fetch&&(window.fetch=function(){var t=o(n.mark(function t(r,o){var i,a,c;return n.wrap(function(t){for(;;)switch(t.prev=t.next){case 0:i=r instanceof Request?r:new Request(r,o),a=0;case 2:if(!(a<e.interceptors.length)){t.next=9;break}return c=e.interceptors[a],t.next=6,Promise.resolve(c(i));case 6:a++,t.next=2;break;case 9:return t.abrupt("return",e.real.call(this,i));case 10:case"end":return t.stop()}},t,this)}));return function(e,n){return t.apply(this,arguments)}}()));}},{key:"add",value:function(e){this.setup(),this.interceptors.push(e);}}]),e}();d(w,"real",void 0),d(w,"hasSetup",!1),d(w,"interceptors",void 0);var k=!1,b=[];function x(){b.forEach(function(e){return e()});}var _=function(){function e(){i(this,e);}return c(e,null,[{key:"route",value:function(e){k||(window.addEventListener("popstate",x,{passive:!0}),window.addEventListener("click",x,{passive:!0}),setTimeout(x),k=!0),b.push(e);}},{key:"create",value:function(e,t){var n=document.createEvent("Event");return n.initEvent(e,t.bubbles||!1,t.cancelable||!0),n.detail=t.detail,n}},{key:"isCrossDomainError",value:function(e){return e instanceof DOMException||"Permission denied"===e.message}}]),e}(),E=function(){function e(){i(this,e);}return c(e,null,[{key:"setup",value:function(t){if(!this.hasSetup||t){this.hasSetup=!0,this.interceptors=[],this.realOpen||(this.realOpen=XMLHttpRequest.prototype.open),this.realSend||(this.realSend=XMLHttpRequest.prototype.send);var n=XMLHttpRequest.prototype;n.open=function(){for(var t=arguments.length,n=new Array(t),r=0;r<t;r++)n[r]=arguments[r];var o=n[1];return this.$url=o,e.realOpen.apply(this,n)},n.send=function(t){for(var n=this,r=[],o=0;o<e.interceptors.length;o++){var i=e.interceptors[o];r.push(i(this,t));}Promise.all(r).then(function(){e.realSend.call(n,t);}).catch(function(e){n.dispatchEvent(_.create("error",{detail:e}));});};}}},{key:"add",value:function(e){this.setup(),this.interceptors.push(e);}}]),e}();d(E,"realOpen",void 0),d(E,"realSend",void 0),d(E,"hasSetup",!1),d(E,"interceptors",void 0);var O=Object.freeze({Fetch:w,XHR:E}),S=function(){function e(t,n,r){i(this,e),d(this,"raw",void 0),d(this,"expiration",void 0),d(this,"type",void 0),this.raw=t,this.expiration=n,this.type=r;}return c(e,[{key:"expired",get:function(){return !this.raw||this.expiration<=Date.now()}}]),e}(),L=function(){function e(t){i(this,e),d(this,"raw",void 0),d(this,"user",void 0),this.raw=t,this.user=e.parse(this.raw);}return c(e,[{key:"expired",get:function(){return !this.user||1e3*this.user.exp<=Date.now()}}],[{key:"parse",value:function(e){try{var t=e.split(".");if(3!==t.length)throw new g({code:"invalid_id_token",message:"ID Token didn't match the desired format. ({header}.{payload}.{validation})"});return JSON.parse(atob(t[1].replace(/-/g,"+").replace(/_/g,"/")))}catch(e){return null}}}]),e}();var A=function(e){if(Array.isArray(e))return e};var T=function(e,t){var n=[],r=!0,o=!1,i=void 0;try{for(var a,c=e[Symbol.iterator]();!(r=(a=c.next()).done)&&(n.push(a.value),!t||n.length!==t);r=!0);}catch(e){o=!0,i=e;}finally{try{r||null==c.return||c.return();}finally{if(o)throw i}}return n};var j=function(){throw new TypeError("Invalid attempt to destructure non-iterable instance")};var $,P,R,U=function(e,t){return A(e)||T(e,t)||j()},q=function(){function e(){i(this,e);}return c(e,null,[{key:"resolve",value:function(e){return $||($=document.implementation.createHTMLDocument("url"),P=$.createElement("base"),R=$.createElement("a"),$.head.appendChild(P)),P.href=window.location.protocol+"//"+window.location.host,R.href=e.replace(/ /g,"%20"),R.href.replace(/\/$/,"")}},{key:"match",value:function(e,t){var n=this;if(t instanceof Array){var r=this.resolve(e),o=D.find(t,function(e){return e instanceof RegExp?!!r.match(e):0===r.indexOf(n.resolve(e))});return !!o}return !0===t}},{key:"parse",value:function(e){var t=e.search,n=e.hash,r=[];t&&(r=r.concat(t.replace("?","").split("&"))),n&&(r=r.concat(n.replace("#","").split("&")));var o={};return D.forEach(r,function(e){var t=e.split("="),n=U(t,2),r=n[0],i=n[1];o[r]=i;}),o}}]),e}(),I={},D=function(){function e(){i(this,e);}return c(e,null,[{key:"includes",value:function(e,t){return -1!==e.indexOf(t)}},{key:"forEach",value:function(e,t){if(Array.isArray(e))for(var n=0;n<e.length;n++)t(e[n],n);else for(var r in e)t(e[r],r);}},{key:"find",value:function(e,t){if(Array.isArray(e))for(var n=0;n<e.length;n++){var r=e[n];if(t(r,n))return r}else for(var o in e){var i=e[o];if(t(i,o))return i}return null}},{key:"assign",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];return this.forEach(n,function(t){for(var n in t)e[n]=t[n];}),e}},{key:"defaults",value:function(e){for(var t=this,n=arguments.length,r=new Array(n>1?n-1:0),o=1;o<n;o++)r[o-1]=arguments[o];return this.forEach(r,function(n){for(var r in n)t.isObject(e[r])&&t.isObject(n[r])?e[r]=t.defaults(e[r],n[r]):void 0===e[r]&&(e[r]=n[r]);}),e}},{key:"isObject",value:function(e){return "object"===s(e)&&!Array.isArray(e)}},{key:"iframe",value:function(){var e=o(n.mark(function e(t){var r,o,i,a;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return r=t.url,o=t.redirectUrl,i=t.visible,(a=document.createElement("iframe")).setAttribute("owner","@salte-auth/salte-auth"),i?(this.assign(a.style,{border:"none",bottom:0,height:"100%",left:0,position:"fixed",right:0,top:0,width:"100%",zIndex:9999,opacity:0,transition:"0.5s opacity"}),setTimeout(function(){return a.style.opacity="1"})):a.style.display="none",a.src=r,document.body.appendChild(a),e.abrupt("return",new Promise(function(e,t){var n=setInterval(function(){try{var r=a.contentWindow.location;if(0!==r.href.indexOf(o))return;var i=q.parse(r);a.parentElement&&a.parentElement.removeChild(a),clearInterval(n),e(i);}catch(e){if(_.isCrossDomainError(e))return;a.parentElement&&a.parentElement.removeChild(a),clearInterval(n),t(e);}});}));case 7:case"end":return e.stop()}},e,this)}));return function(t){return e.apply(this,arguments)}}()},{key:"debounce",value:function(e,t,n){clearTimeout(I[e]),I[e]=window.setTimeout(function(){delete I[e],t();},n);}}]),e}(),M=function(){function e(t,n){i(this,e),d(this,"name",void 0),d(this,"level",void 0),this.name=t,this.level="string"==typeof n?this.toLevel(n):n;}return c(e,[{key:"trace",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];this.log.apply(this,["trace",e].concat(n));}},{key:"info",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];this.log.apply(this,["info",e].concat(n));}},{key:"warn",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];this.log.apply(this,["warn",e].concat(n));}},{key:"error",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];this.log.apply(this,["error",e].concat(n));}},{key:"log",value:function(e,t){if(this.enabled(e)){for(var n,r=arguments.length,o=new Array(r>2?r-2:0),i=2;i<r;i++)o[i-2]=arguments[i];(n=console).log.apply(n,["".concat(e,": ").concat(t)].concat(o));}}},{key:"enabled",value:function(e){return !1!==this.level&&(!0===this.level||this.level<=this.toLevel(e))}},{key:"toLevel",value:function(t){return D.find(e.levels,function(e,n){return n===t})}}]),e}();d(M,"levels",{trace:0,info:1,warn:2,error:3});var N=Object.freeze({Interceptors:O,AccessToken:S,IDToken:L,Common:D,Events:_,URL:q,Logger:M}),C=function(e){function t(e){var n;return i(this,t),(n=l(this,f(t).call(this,e))).config=D.defaults(n.config,{storage:"session"}),n}return p(t,m),c(t,[{key:"get",value:function(e){var t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:null;return this.storage.getItem(this.key(e))||t}},{key:"set",value:function(e,t){D.includes([void 0,null],t)?this.clear(e):this.storage.setItem(this.key(e),t);}},{key:"clear",value:function(e){this.storage.removeItem(this.key(e));}},{key:"reset",value:function(){var e=this.key("");for(var t in localStorage)0===t.indexOf(e)&&localStorage.removeItem(t);for(var n in sessionStorage)0===n.indexOf(e)&&sessionStorage.removeItem(n);}},{key:"key",value:function(e){return "salte.auth.".concat(e)}},{key:"storage",get:function(){var e=this.config&&this.config.storage;switch(e){case"local":return localStorage;case"session":return sessionStorage}throw new g({code:"invalid_storage",message:"Storage doesn't exist for the given value. (".concat(e,")")})}}]),t}(),F=function(e){function t(){var e,n;i(this,t);for(var r=arguments.length,o=new Array(r),a=0;a<r;a++)o[a]=arguments[a];return n=l(this,(e=f(t)).call.apply(e,[this].concat(o))),d(u(n),"listeners",new Map),n}return p(t,C),c(t,[{key:"on",value:function(e,t){this.listeners.has(e)||this.listeners.set(e,[]),this.listeners.get(e).push(t);}},{key:"off",value:function(e,t){if(this.listeners.has(e)){var n=this.listeners.get(e);if(n.length){var r=n.indexOf(t);-1!==r&&n.splice(r,1);}}}},{key:"emit",value:function(e){for(var t=arguments.length,n=new Array(t>1?t-1:0),r=1;r<t;r++)n[r-1]=arguments[r];if(this.listeners.has(e)){var o=this.listeners.get(e);D.forEach(o,function(e){return e.apply(void 0,n)});}}}]),t}(),H=function(e){function t(e){var n;return i(this,t),(n=l(this,f(t).call(this,e))).config=D.defaults(n.config,{redirectUrl:location.origin,level:"warn"}),n}return p(t,F),c(t,[{key:"redirectUrl",value:function(e){return "string"==typeof this.config.redirectUrl?this.config.redirectUrl:this.config.redirectUrl[e]}}]),t}(),G=self.crypto||self.msCrypto,z=function(e){e=e||21;for(var t="",n=G.getRandomValues(new Uint8Array(e));0<e--;)t+="Uint8ArdomValuesObj012345679BCDEFGHIJKLMNPQRSTWXYZ_cfghkpqvwxyz-"[63&n[e]];return t},X=function(e){function t(e){var n;return i(this,t),n=l(this,f(t).call(this,e)),d(u(n),"logger",void 0),n.config=D.defaults(n.config,{validation:!0,level:"warn"}),n.logger=new M("@salte-auth/salte-auth:providers/".concat(n.$name),n.config.level),n}return p(t,H),c(t,[{key:"validation",value:function(e){return "object"===s(this.config.validation)?!0===this.config.validation[e]:!0===this.config.validation}},{key:"key",value:function(e){function t(t){return e.apply(this,arguments)}return t.toString=function(){return e.toString()},t}(function(e){return "salte.auth.provider.".concat(this.$name,".").concat(e)})},{key:"url",value:function(e){var t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},n=e;return D.forEach(t,function(e,t){D.includes([void 0,null,""],e)||(n+="".concat(-1===n.indexOf("?")?"?":"&").concat(t,"=").concat(encodeURIComponent(e)));}),n}},{key:"$name",get:function(){return this.config.name||this.name}}]),t}(),B=function(e){function t(e){return i(this,t),l(this,f(t).call(this,e))}return p(t,X),c(t,[{key:"connected",value:function(){this.required("responseType");}},{key:"secure",value:function(){var e=o(n.mark(function e(t){var r;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:if(!D.includes(["token"],this.config.responseType)){e.next=14;break}if(!(r=this.accessToken()).expired){e.next=4;break}return e.abrupt("return",this.$login());case 4:if(!t){e.next=14;break}if(!(t instanceof Request)){e.next=9;break}t.headers.set("Authorization","Bearer ".concat(r.raw)),e.next=14;break;case 9:if(!(t instanceof XMLHttpRequest)){e.next=13;break}t.setRequestHeader("Authorization","Bearer ".concat(r.raw)),e.next=14;break;case 13:throw new g({code:"unknown_request",message:"Unknown request type. (".concat(t,")")});case 14:return e.abrupt("return",!0);case 15:case"end":return e.stop()}},e,this)}));return function(t){return e.apply(this,arguments)}}()},{key:"$validate",value:function(e){try{if(e.error)throw new g({code:e.error,message:"".concat(e.error_description?e.error_description:e.error).concat(e.error_uri?" (".concat(e.error_uri,")"):"")});var t=e.code,n=e.access_token,r=e.state,o=e.expires_in,i=e.token_type;if(this.validation("state")&&this.get("state")!==r)throw new g({code:"invalid_state",message:"State provided by identity provider did not match local state."});var a=this.get("response-type","").split(" ");if(D.includes(a,"code")){if(!t)throw new g({code:"invalid_code",message:"Expected a code to be returned by the Provider."})}else if(D.includes(a,"token")&&!n)throw new g({code:"invalid_access_token",message:"Expected an access token to be returned by the Provider."});t?(this.set("code.raw",t),this.clear("access-token.raw"),this.clear("access-token.expiration"),this.clear("access-token.type")):n&&(this.set("access-token.raw",n),this.set("access-token.expiration",Date.now()+1e3*Number(o)),this.set("access-token.type",i),this.clear("code.raw"));}finally{this.clear("state");}}},{key:"validate",value:function(e){try{this.$validate(e);}catch(e){throw this.emit("login",e),e}this.emit("login",null,this.code||this.accessToken());}},{key:"accessToken",value:function(){var e=this.get("access-token.expiration");return new S(this.get("access-token.raw"),D.includes([void 0,null],e)?null:Number(e),this.get("access-token.type"))}},{key:"$login",value:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},t="".concat(this.$name,"-state-").concat(z()),n=e.responseType||this.config.responseType;return this.set("state",t),this.set("response-type",n),this.url(this.login,{client_id:this.config.clientID,response_type:n,redirect_uri:this.config.redirectUrl,scope:this.config.scope,state:t})}},{key:"code",get:function(){return this.get("code.raw")}},{key:"logout",get:function(){throw new g({code:"logout_not_supported",message:"OAuth 2.0 doesn't support logout!"})}}]),t}();var Y=function(e,t){for(;!Object.prototype.hasOwnProperty.call(e,t)&&null!==(e=f(e)););return e},J=t(function(e){function t(n,r,o){return "undefined"!=typeof Reflect&&Reflect.get?e.exports=t=Reflect.get:e.exports=t=function(e,t,n){var r=Y(e,t);if(r){var o=Object.getOwnPropertyDescriptor(r,t);return o.get?o.get.call(n):o.value}},t(n,r,o||n)}e.exports=t;}),V=function(e){function t(e){var n;return i(this,t),(n=l(this,f(t).call(this,e))).config.renewal="object"===s(n.config.renewal)?n.config.renewal:{type:n.config.renewal},n.config=D.defaults(n.config,{responseType:"id_token",scope:"openid",renewal:{type:"auto",buffer:6e4}}),n}return p(t,B),c(t,[{key:"secure",value:function(){var e=o(n.mark(function e(t){var r;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:if(!D.includes(["id_token","id_token token","token"],this.config.responseType)){e.next=18;break}if(!this.idToken().expired){e.next=3;break}return e.abrupt("return",this.$login());case 3:if(!this.accessToken().expired){e.next=8;break}return e.next=6,D.iframe({redirectUrl:this.redirectUrl("login"),url:this.$login({prompt:"none",responseType:"token"})});case 6:r=e.sent,this.validate(r);case 8:if(!t){e.next=18;break}if(!(t instanceof Request)){e.next=13;break}t.headers.set("Authorization","Bearer ".concat(this.accessToken().raw)),e.next=18;break;case 13:if(!(t instanceof XMLHttpRequest)){e.next=17;break}t.setRequestHeader("Authorization","Bearer ".concat(this.accessToken().raw)),e.next=18;break;case 17:throw new g({code:"unknown_request",message:"Unknown request type. (".concat(t,")")});case 18:return e.abrupt("return",!0);case 19:case"end":return e.stop()}},e,this)}));return function(t){return e.apply(this,arguments)}}()},{key:"$validate",value:function(e){try{J(f(t.prototype),"$validate",this).call(this,e);var n=this.get("response-type","").split(" ");if(D.includes(n,"id_token")){var r=e.id_token,o=L.parse(r);if(!o)throw new g({code:"invalid_id_token",message:"Failed to parse user information due to invalid id token."});if(this.validation("nonce")&&this.get("nonce")!==o.nonce)throw new g({code:"invalid_nonce",message:"Nonce provided by identity provider did not match the local nonce."});this.set("id-token.raw",r);}}finally{this.clear("nonce");}}},{key:"validate",value:function(e){try{this.$validate(e);}catch(e){throw this.emit("login",e),e}this.emit("login",null,this.code||this.idToken());}},{key:"idToken",value:function(){return new L(this.get("id-token.raw"))}},{key:"$login",value:function(e){var n="".concat(this.$name,"-nonce-").concat(z());return this.set("nonce",n),this.url(J(f(t.prototype),"$login",this).call(this,e),{prompt:e&&e.prompt,nonce:n})}}]),t}(),W=function(e){function t(e){var n;return i(this,t),(n=l(this,f(t).call(this,e))).required("login"),n}return p(t,B),c(t,[{key:"name",get:function(){return "generic.oauth2"}},{key:"login",get:function(){return this.config.login.apply(this)}}]),t}(),K=function(e){function t(e){var n;return i(this,t),(n=l(this,f(t).call(this,e))).required("login","logout"),n}return p(t,V),c(t,[{key:"name",get:function(){return "generic.openid"}},{key:"login",get:function(){return this.config.login.apply(this)}},{key:"logout",get:function(){return this.config.logout.apply(this)}}]),t}(),Q=Object.freeze({OAuth2:W,OpenID:K}),Z=function(e){function t(e){var n;return i(this,t),n=l(this,f(t).call(this,e)),d(u(n),"logger",void 0),n.config=D.defaults(n.config,{level:"warn"}),n.logger=new M("@salte-auth/salte-auth:handlers/".concat(n.$name),n.config.level),n}return p(t,C),c(t,[{key:"key",value:function(e){return "salte.auth.handler.".concat(this.$name,".").concat(e)}},{key:"navigate",value:function(e){location.href=e;}},{key:"$name",get:function(){return this.config.name||this.name}}]),t}(),ee=function(e){function t(e){var r;i(this,t),r=l(this,f(t).call(this,e)),d(u(r),"logger",void 0),r.required("providers","handlers"),r.config=D.defaults(r.config,{validation:!0,level:"warn"}),r.logger=new M("@salte-auth/salte-auth:core",r.config.level),D.forEach(r.config.providers,function(e){e.connected&&e.connected(),e.on("login",function(e,t){r.emit("login",e,t);}),e.on("logout",function(e){r.emit("logout",e);});});var a=r.get("provider"),c=a?r.provider(a):null;return D.forEach(r.config.handlers,function(e){e.connected&&e.connected({action:r.get("action"),handler:r.get("handler"),provider:c});}),r.clear("action"),r.clear("provider"),r.clear("handler"),w.add(function(){var e=o(n.mark(function e(t){var o,i;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:o=0;case 1:if(!(o<r.config.providers.length)){e.next=11;break}if(i=r.config.providers[o],!q.match(t.url,i.config.endpoints)){e.next=8;break}if(e.t0=i.secure,!e.t0){e.next=8;break}return e.next=8,i.secure(t);case 8:o++,e.next=1;break;case 11:case"end":return e.stop()}},e)}));return function(t){return e.apply(this,arguments)}}()),E.add(function(){var e=o(n.mark(function e(t){var o,i;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:o=0;case 1:if(!(o<r.config.providers.length)){e.next=11;break}if(i=r.config.providers[o],!q.match(t.$url,i.config.endpoints)){e.next=8;break}if(e.t0=i.secure,!e.t0){e.next=8;break}return e.next=8,i.secure(t);case 8:o++,e.next=1;break;case 11:case"end":return e.stop()}},e)}));return function(t){return e.apply(this,arguments)}}()),_.route(o(n.mark(function e(){var t,o,i,a,c;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:e.prev=0,t=r.handler(),o=0;case 3:if(!(o<r.config.providers.length)){e.next=29;break}if(i=r.config.providers[o],!q.match(location.href,i.config.routes)){e.next=26;break}a=null;case 7:if(!0===a){e.next=26;break}return e.next=10,i.secure();case 10:if("string"!=typeof(a=e.sent)){e.next=24;break}if(t.auto){e.next=14;break}throw new g({code:"auto_unsupported",message:"The default handler doesn't support automatic authentication! (".concat(t.$name,")")});case 14:return r.set("action","login"),r.set("provider",i.$name),r.set("handler",t.$name),e.next=19,t.open({redirectUrl:i.redirectUrl("login"),url:a});case 19:c=e.sent,i.validate(c),r.clear("action"),r.clear("provider"),r.clear("handler");case 24:e.next=7;break;case 26:o++,e.next=3;break;case 29:e.next=37;break;case 31:throw e.prev=31,e.t0=e.catch(0),r.clear("action"),r.clear("provider"),r.clear("handler"),e.t0;case 37:case"end":return e.stop()}},e,null,[[0,31]])}))),r}return p(t,H),c(t,[{key:"login",value:function(){var e=o(n.mark(function e(t){var r,o,i;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return t="string"==typeof t?{provider:t}:t,e.prev=1,r=this.provider(t.provider),o=this.handler(t.handler),this.set("action","login"),this.set("provider",r.$name),this.set("handler",o.$name),e.next=9,o.open({redirectUrl:r.redirectUrl("login"),url:r.$login()});case 9:i=e.sent,r.validate(i);case 11:return e.prev=11,this.clear("action"),this.clear("provider"),this.clear("handler"),e.finish(11);case 16:case"end":return e.stop()}},e,this,[[1,,11,16]])}));return function(t){return e.apply(this,arguments)}}()},{key:"logout",value:function(){var e=o(n.mark(function e(t){var r,o;return n.wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return t="string"==typeof t?{provider:t}:t,r=this.provider(t.provider),e.prev=2,o=this.handler(t.handler),this.set("action","logout"),this.set("provider",r.$name),this.set("handler",o.$name),e.next=9,o.open({redirectUrl:r.redirectUrl("logout"),url:r.logout});case 9:r.reset(),r.emit("logout"),e.next=17;break;case 13:throw e.prev=13,e.t0=e.catch(2),r.emit("logout",e.t0),e.t0;case 17:return e.prev=17,this.clear("action"),this.clear("provider"),this.clear("handler"),e.finish(17);case 22:case"end":return e.stop()}},e,this,[[2,13,17,22]])}));return function(t){return e.apply(this,arguments)}}()},{key:"provider",value:function(e){var t=D.find(this.config.providers,function(t){return t.$name===e});if(!t)throw new g({code:"invalid_provider",message:"Unable to locate provider with the given name. (".concat(e,")")});return t}},{key:"handler",value:function(e){var t=void 0===e?D.find(this.config.handlers,function(e){return !!e.config.default}):D.find(this.config.handlers,function(t){return t.$name===e});if(!t)throw new g({code:"invalid_handler",message:"Unable to locate handler with the given name. (".concat(e,")")});return t}}]),t}();e.Generic=Q,e.Handler=Z,e.OAuth2Provider=B,e.OpenIDProvider=V,e.SalteAuth=ee,e.SalteAuthError=g,e.Utils=N,Object.defineProperty(e,"__esModule",{value:!0});});

	});

	unwrapExports(salteAuth_min);
	var salteAuth_min_1 = salteAuth_min.SalteAuth;
	var salteAuth_min_2 = salteAuth_min.SalteAuthError;
	var salteAuth_min_3 = salteAuth_min.OAuth2Provider;
	var salteAuth_min_4 = salteAuth_min.OpenIDProvider;
	var salteAuth_min_5 = salteAuth_min.Handler;
	var salteAuth_min_6 = salteAuth_min.Utils;
	var salteAuth_min_7 = salteAuth_min.Generic;

	var runtime_1 = createCommonjsModule(function (module) {
	/**
	 * Copyright (c) 2014-present, Facebook, Inc.
	 *
	 * This source code is licensed under the MIT license found in the
	 * LICENSE file in the root directory of this source tree.
	 */

	var runtime = (function (exports) {

	  var Op = Object.prototype;
	  var hasOwn = Op.hasOwnProperty;
	  var undefined$1; // More compressible than void 0.
	  var $Symbol = typeof Symbol === "function" ? Symbol : {};
	  var iteratorSymbol = $Symbol.iterator || "@@iterator";
	  var asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator";
	  var toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag";

	  function wrap(innerFn, outerFn, self, tryLocsList) {
	    // If outerFn provided and outerFn.prototype is a Generator, then outerFn.prototype instanceof Generator.
	    var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator;
	    var generator = Object.create(protoGenerator.prototype);
	    var context = new Context(tryLocsList || []);

	    // The ._invoke method unifies the implementations of the .next,
	    // .throw, and .return methods.
	    generator._invoke = makeInvokeMethod(innerFn, self, context);

	    return generator;
	  }
	  exports.wrap = wrap;

	  // Try/catch helper to minimize deoptimizations. Returns a completion
	  // record like context.tryEntries[i].completion. This interface could
	  // have been (and was previously) designed to take a closure to be
	  // invoked without arguments, but in all the cases we care about we
	  // already have an existing method we want to call, so there's no need
	  // to create a new function object. We can even get away with assuming
	  // the method takes exactly one argument, since that happens to be true
	  // in every case, so we don't have to touch the arguments object. The
	  // only additional allocation required is the completion record, which
	  // has a stable shape and so hopefully should be cheap to allocate.
	  function tryCatch(fn, obj, arg) {
	    try {
	      return { type: "normal", arg: fn.call(obj, arg) };
	    } catch (err) {
	      return { type: "throw", arg: err };
	    }
	  }

	  var GenStateSuspendedStart = "suspendedStart";
	  var GenStateSuspendedYield = "suspendedYield";
	  var GenStateExecuting = "executing";
	  var GenStateCompleted = "completed";

	  // Returning this object from the innerFn has the same effect as
	  // breaking out of the dispatch switch statement.
	  var ContinueSentinel = {};

	  // Dummy constructor functions that we use as the .constructor and
	  // .constructor.prototype properties for functions that return Generator
	  // objects. For full spec compliance, you may wish to configure your
	  // minifier not to mangle the names of these two functions.
	  function Generator() {}
	  function GeneratorFunction() {}
	  function GeneratorFunctionPrototype() {}

	  // This is a polyfill for %IteratorPrototype% for environments that
	  // don't natively support it.
	  var IteratorPrototype = {};
	  IteratorPrototype[iteratorSymbol] = function () {
	    return this;
	  };

	  var getProto = Object.getPrototypeOf;
	  var NativeIteratorPrototype = getProto && getProto(getProto(values([])));
	  if (NativeIteratorPrototype &&
	      NativeIteratorPrototype !== Op &&
	      hasOwn.call(NativeIteratorPrototype, iteratorSymbol)) {
	    // This environment has a native %IteratorPrototype%; use it instead
	    // of the polyfill.
	    IteratorPrototype = NativeIteratorPrototype;
	  }

	  var Gp = GeneratorFunctionPrototype.prototype =
	    Generator.prototype = Object.create(IteratorPrototype);
	  GeneratorFunction.prototype = Gp.constructor = GeneratorFunctionPrototype;
	  GeneratorFunctionPrototype.constructor = GeneratorFunction;
	  GeneratorFunctionPrototype[toStringTagSymbol] =
	    GeneratorFunction.displayName = "GeneratorFunction";

	  // Helper for defining the .next, .throw, and .return methods of the
	  // Iterator interface in terms of a single ._invoke method.
	  function defineIteratorMethods(prototype) {
	    ["next", "throw", "return"].forEach(function(method) {
	      prototype[method] = function(arg) {
	        return this._invoke(method, arg);
	      };
	    });
	  }

	  exports.isGeneratorFunction = function(genFun) {
	    var ctor = typeof genFun === "function" && genFun.constructor;
	    return ctor
	      ? ctor === GeneratorFunction ||
	        // For the native GeneratorFunction constructor, the best we can
	        // do is to check its .name property.
	        (ctor.displayName || ctor.name) === "GeneratorFunction"
	      : false;
	  };

	  exports.mark = function(genFun) {
	    if (Object.setPrototypeOf) {
	      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
	    } else {
	      genFun.__proto__ = GeneratorFunctionPrototype;
	      if (!(toStringTagSymbol in genFun)) {
	        genFun[toStringTagSymbol] = "GeneratorFunction";
	      }
	    }
	    genFun.prototype = Object.create(Gp);
	    return genFun;
	  };

	  // Within the body of any async function, `await x` is transformed to
	  // `yield regeneratorRuntime.awrap(x)`, so that the runtime can test
	  // `hasOwn.call(value, "__await")` to determine if the yielded value is
	  // meant to be awaited.
	  exports.awrap = function(arg) {
	    return { __await: arg };
	  };

	  function AsyncIterator(generator) {
	    function invoke(method, arg, resolve, reject) {
	      var record = tryCatch(generator[method], generator, arg);
	      if (record.type === "throw") {
	        reject(record.arg);
	      } else {
	        var result = record.arg;
	        var value = result.value;
	        if (value &&
	            typeof value === "object" &&
	            hasOwn.call(value, "__await")) {
	          return Promise.resolve(value.__await).then(function(value) {
	            invoke("next", value, resolve, reject);
	          }, function(err) {
	            invoke("throw", err, resolve, reject);
	          });
	        }

	        return Promise.resolve(value).then(function(unwrapped) {
	          // When a yielded Promise is resolved, its final value becomes
	          // the .value of the Promise<{value,done}> result for the
	          // current iteration.
	          result.value = unwrapped;
	          resolve(result);
	        }, function(error) {
	          // If a rejected Promise was yielded, throw the rejection back
	          // into the async generator function so it can be handled there.
	          return invoke("throw", error, resolve, reject);
	        });
	      }
	    }

	    var previousPromise;

	    function enqueue(method, arg) {
	      function callInvokeWithMethodAndArg() {
	        return new Promise(function(resolve, reject) {
	          invoke(method, arg, resolve, reject);
	        });
	      }

	      return previousPromise =
	        // If enqueue has been called before, then we want to wait until
	        // all previous Promises have been resolved before calling invoke,
	        // so that results are always delivered in the correct order. If
	        // enqueue has not been called before, then it is important to
	        // call invoke immediately, without waiting on a callback to fire,
	        // so that the async generator function has the opportunity to do
	        // any necessary setup in a predictable way. This predictability
	        // is why the Promise constructor synchronously invokes its
	        // executor callback, and why async functions synchronously
	        // execute code before the first await. Since we implement simple
	        // async functions in terms of async generators, it is especially
	        // important to get this right, even though it requires care.
	        previousPromise ? previousPromise.then(
	          callInvokeWithMethodAndArg,
	          // Avoid propagating failures to Promises returned by later
	          // invocations of the iterator.
	          callInvokeWithMethodAndArg
	        ) : callInvokeWithMethodAndArg();
	    }

	    // Define the unified helper method that is used to implement .next,
	    // .throw, and .return (see defineIteratorMethods).
	    this._invoke = enqueue;
	  }

	  defineIteratorMethods(AsyncIterator.prototype);
	  AsyncIterator.prototype[asyncIteratorSymbol] = function () {
	    return this;
	  };
	  exports.AsyncIterator = AsyncIterator;

	  // Note that simple async functions are implemented on top of
	  // AsyncIterator objects; they just return a Promise for the value of
	  // the final result produced by the iterator.
	  exports.async = function(innerFn, outerFn, self, tryLocsList) {
	    var iter = new AsyncIterator(
	      wrap(innerFn, outerFn, self, tryLocsList)
	    );

	    return exports.isGeneratorFunction(outerFn)
	      ? iter // If outerFn is a generator, return the full iterator.
	      : iter.next().then(function(result) {
	          return result.done ? result.value : iter.next();
	        });
	  };

	  function makeInvokeMethod(innerFn, self, context) {
	    var state = GenStateSuspendedStart;

	    return function invoke(method, arg) {
	      if (state === GenStateExecuting) {
	        throw new Error("Generator is already running");
	      }

	      if (state === GenStateCompleted) {
	        if (method === "throw") {
	          throw arg;
	        }

	        // Be forgiving, per 25.3.3.3.3 of the spec:
	        // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
	        return doneResult();
	      }

	      context.method = method;
	      context.arg = arg;

	      while (true) {
	        var delegate = context.delegate;
	        if (delegate) {
	          var delegateResult = maybeInvokeDelegate(delegate, context);
	          if (delegateResult) {
	            if (delegateResult === ContinueSentinel) continue;
	            return delegateResult;
	          }
	        }

	        if (context.method === "next") {
	          // Setting context._sent for legacy support of Babel's
	          // function.sent implementation.
	          context.sent = context._sent = context.arg;

	        } else if (context.method === "throw") {
	          if (state === GenStateSuspendedStart) {
	            state = GenStateCompleted;
	            throw context.arg;
	          }

	          context.dispatchException(context.arg);

	        } else if (context.method === "return") {
	          context.abrupt("return", context.arg);
	        }

	        state = GenStateExecuting;

	        var record = tryCatch(innerFn, self, context);
	        if (record.type === "normal") {
	          // If an exception is thrown from innerFn, we leave state ===
	          // GenStateExecuting and loop back for another invocation.
	          state = context.done
	            ? GenStateCompleted
	            : GenStateSuspendedYield;

	          if (record.arg === ContinueSentinel) {
	            continue;
	          }

	          return {
	            value: record.arg,
	            done: context.done
	          };

	        } else if (record.type === "throw") {
	          state = GenStateCompleted;
	          // Dispatch the exception by looping back around to the
	          // context.dispatchException(context.arg) call above.
	          context.method = "throw";
	          context.arg = record.arg;
	        }
	      }
	    };
	  }

	  // Call delegate.iterator[context.method](context.arg) and handle the
	  // result, either by returning a { value, done } result from the
	  // delegate iterator, or by modifying context.method and context.arg,
	  // setting context.delegate to null, and returning the ContinueSentinel.
	  function maybeInvokeDelegate(delegate, context) {
	    var method = delegate.iterator[context.method];
	    if (method === undefined$1) {
	      // A .throw or .return when the delegate iterator has no .throw
	      // method always terminates the yield* loop.
	      context.delegate = null;

	      if (context.method === "throw") {
	        // Note: ["return"] must be used for ES3 parsing compatibility.
	        if (delegate.iterator["return"]) {
	          // If the delegate iterator has a return method, give it a
	          // chance to clean up.
	          context.method = "return";
	          context.arg = undefined$1;
	          maybeInvokeDelegate(delegate, context);

	          if (context.method === "throw") {
	            // If maybeInvokeDelegate(context) changed context.method from
	            // "return" to "throw", let that override the TypeError below.
	            return ContinueSentinel;
	          }
	        }

	        context.method = "throw";
	        context.arg = new TypeError(
	          "The iterator does not provide a 'throw' method");
	      }

	      return ContinueSentinel;
	    }

	    var record = tryCatch(method, delegate.iterator, context.arg);

	    if (record.type === "throw") {
	      context.method = "throw";
	      context.arg = record.arg;
	      context.delegate = null;
	      return ContinueSentinel;
	    }

	    var info = record.arg;

	    if (! info) {
	      context.method = "throw";
	      context.arg = new TypeError("iterator result is not an object");
	      context.delegate = null;
	      return ContinueSentinel;
	    }

	    if (info.done) {
	      // Assign the result of the finished delegate to the temporary
	      // variable specified by delegate.resultName (see delegateYield).
	      context[delegate.resultName] = info.value;

	      // Resume execution at the desired location (see delegateYield).
	      context.next = delegate.nextLoc;

	      // If context.method was "throw" but the delegate handled the
	      // exception, let the outer generator proceed normally. If
	      // context.method was "next", forget context.arg since it has been
	      // "consumed" by the delegate iterator. If context.method was
	      // "return", allow the original .return call to continue in the
	      // outer generator.
	      if (context.method !== "return") {
	        context.method = "next";
	        context.arg = undefined$1;
	      }

	    } else {
	      // Re-yield the result returned by the delegate method.
	      return info;
	    }

	    // The delegate iterator is finished, so forget it and continue with
	    // the outer generator.
	    context.delegate = null;
	    return ContinueSentinel;
	  }

	  // Define Generator.prototype.{next,throw,return} in terms of the
	  // unified ._invoke helper method.
	  defineIteratorMethods(Gp);

	  Gp[toStringTagSymbol] = "Generator";

	  // A Generator should always return itself as the iterator object when the
	  // @@iterator function is called on it. Some browsers' implementations of the
	  // iterator prototype chain incorrectly implement this, causing the Generator
	  // object to not be returned from this call. This ensures that doesn't happen.
	  // See https://github.com/facebook/regenerator/issues/274 for more details.
	  Gp[iteratorSymbol] = function() {
	    return this;
	  };

	  Gp.toString = function() {
	    return "[object Generator]";
	  };

	  function pushTryEntry(locs) {
	    var entry = { tryLoc: locs[0] };

	    if (1 in locs) {
	      entry.catchLoc = locs[1];
	    }

	    if (2 in locs) {
	      entry.finallyLoc = locs[2];
	      entry.afterLoc = locs[3];
	    }

	    this.tryEntries.push(entry);
	  }

	  function resetTryEntry(entry) {
	    var record = entry.completion || {};
	    record.type = "normal";
	    delete record.arg;
	    entry.completion = record;
	  }

	  function Context(tryLocsList) {
	    // The root entry object (effectively a try statement without a catch
	    // or a finally block) gives us a place to store values thrown from
	    // locations where there is no enclosing try statement.
	    this.tryEntries = [{ tryLoc: "root" }];
	    tryLocsList.forEach(pushTryEntry, this);
	    this.reset(true);
	  }

	  exports.keys = function(object) {
	    var keys = [];
	    for (var key in object) {
	      keys.push(key);
	    }
	    keys.reverse();

	    // Rather than returning an object with a next method, we keep
	    // things simple and return the next function itself.
	    return function next() {
	      while (keys.length) {
	        var key = keys.pop();
	        if (key in object) {
	          next.value = key;
	          next.done = false;
	          return next;
	        }
	      }

	      // To avoid creating an additional object, we just hang the .value
	      // and .done properties off the next function object itself. This
	      // also ensures that the minifier will not anonymize the function.
	      next.done = true;
	      return next;
	    };
	  };

	  function values(iterable) {
	    if (iterable) {
	      var iteratorMethod = iterable[iteratorSymbol];
	      if (iteratorMethod) {
	        return iteratorMethod.call(iterable);
	      }

	      if (typeof iterable.next === "function") {
	        return iterable;
	      }

	      if (!isNaN(iterable.length)) {
	        var i = -1, next = function next() {
	          while (++i < iterable.length) {
	            if (hasOwn.call(iterable, i)) {
	              next.value = iterable[i];
	              next.done = false;
	              return next;
	            }
	          }

	          next.value = undefined$1;
	          next.done = true;

	          return next;
	        };

	        return next.next = next;
	      }
	    }

	    // Return an iterator with no values.
	    return { next: doneResult };
	  }
	  exports.values = values;

	  function doneResult() {
	    return { value: undefined$1, done: true };
	  }

	  Context.prototype = {
	    constructor: Context,

	    reset: function(skipTempReset) {
	      this.prev = 0;
	      this.next = 0;
	      // Resetting context._sent for legacy support of Babel's
	      // function.sent implementation.
	      this.sent = this._sent = undefined$1;
	      this.done = false;
	      this.delegate = null;

	      this.method = "next";
	      this.arg = undefined$1;

	      this.tryEntries.forEach(resetTryEntry);

	      if (!skipTempReset) {
	        for (var name in this) {
	          // Not sure about the optimal order of these conditions:
	          if (name.charAt(0) === "t" &&
	              hasOwn.call(this, name) &&
	              !isNaN(+name.slice(1))) {
	            this[name] = undefined$1;
	          }
	        }
	      }
	    },

	    stop: function() {
	      this.done = true;

	      var rootEntry = this.tryEntries[0];
	      var rootRecord = rootEntry.completion;
	      if (rootRecord.type === "throw") {
	        throw rootRecord.arg;
	      }

	      return this.rval;
	    },

	    dispatchException: function(exception) {
	      if (this.done) {
	        throw exception;
	      }

	      var context = this;
	      function handle(loc, caught) {
	        record.type = "throw";
	        record.arg = exception;
	        context.next = loc;

	        if (caught) {
	          // If the dispatched exception was caught by a catch block,
	          // then let that catch block handle the exception normally.
	          context.method = "next";
	          context.arg = undefined$1;
	        }

	        return !! caught;
	      }

	      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
	        var entry = this.tryEntries[i];
	        var record = entry.completion;

	        if (entry.tryLoc === "root") {
	          // Exception thrown outside of any try block that could handle
	          // it, so set the completion value of the entire function to
	          // throw the exception.
	          return handle("end");
	        }

	        if (entry.tryLoc <= this.prev) {
	          var hasCatch = hasOwn.call(entry, "catchLoc");
	          var hasFinally = hasOwn.call(entry, "finallyLoc");

	          if (hasCatch && hasFinally) {
	            if (this.prev < entry.catchLoc) {
	              return handle(entry.catchLoc, true);
	            } else if (this.prev < entry.finallyLoc) {
	              return handle(entry.finallyLoc);
	            }

	          } else if (hasCatch) {
	            if (this.prev < entry.catchLoc) {
	              return handle(entry.catchLoc, true);
	            }

	          } else if (hasFinally) {
	            if (this.prev < entry.finallyLoc) {
	              return handle(entry.finallyLoc);
	            }

	          } else {
	            throw new Error("try statement without catch or finally");
	          }
	        }
	      }
	    },

	    abrupt: function(type, arg) {
	      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
	        var entry = this.tryEntries[i];
	        if (entry.tryLoc <= this.prev &&
	            hasOwn.call(entry, "finallyLoc") &&
	            this.prev < entry.finallyLoc) {
	          var finallyEntry = entry;
	          break;
	        }
	      }

	      if (finallyEntry &&
	          (type === "break" ||
	           type === "continue") &&
	          finallyEntry.tryLoc <= arg &&
	          arg <= finallyEntry.finallyLoc) {
	        // Ignore the finally entry if control is not jumping to a
	        // location outside the try/catch block.
	        finallyEntry = null;
	      }

	      var record = finallyEntry ? finallyEntry.completion : {};
	      record.type = type;
	      record.arg = arg;

	      if (finallyEntry) {
	        this.method = "next";
	        this.next = finallyEntry.finallyLoc;
	        return ContinueSentinel;
	      }

	      return this.complete(record);
	    },

	    complete: function(record, afterLoc) {
	      if (record.type === "throw") {
	        throw record.arg;
	      }

	      if (record.type === "break" ||
	          record.type === "continue") {
	        this.next = record.arg;
	      } else if (record.type === "return") {
	        this.rval = this.arg = record.arg;
	        this.method = "return";
	        this.next = "end";
	      } else if (record.type === "normal" && afterLoc) {
	        this.next = afterLoc;
	      }

	      return ContinueSentinel;
	    },

	    finish: function(finallyLoc) {
	      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
	        var entry = this.tryEntries[i];
	        if (entry.finallyLoc === finallyLoc) {
	          this.complete(entry.completion, entry.afterLoc);
	          resetTryEntry(entry);
	          return ContinueSentinel;
	        }
	      }
	    },

	    "catch": function(tryLoc) {
	      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
	        var entry = this.tryEntries[i];
	        if (entry.tryLoc === tryLoc) {
	          var record = entry.completion;
	          if (record.type === "throw") {
	            var thrown = record.arg;
	            resetTryEntry(entry);
	          }
	          return thrown;
	        }
	      }

	      // The context.catch method must only be called with a location
	      // argument that corresponds to a known catch block.
	      throw new Error("illegal catch attempt");
	    },

	    delegateYield: function(iterable, resultName, nextLoc) {
	      this.delegate = {
	        iterator: values(iterable),
	        resultName: resultName,
	        nextLoc: nextLoc
	      };

	      if (this.method === "next") {
	        // Deliberately forget the last sent value so that we don't
	        // accidentally pass it on to the delegate.
	        this.arg = undefined$1;
	      }

	      return ContinueSentinel;
	    }
	  };

	  // Regardless of whether this script is executing as a CommonJS module
	  // or not, return the runtime object so that we can declare the variable
	  // regeneratorRuntime in the outer scope, which allows this module to be
	  // injected easily by `bin/regenerator --include-runtime script.js`.
	  return exports;

	}(
	  // If this script is executing as a CommonJS module, use module.exports
	  // as the regeneratorRuntime namespace. Otherwise create a new empty
	  // object. Either way, the resulting object will be used to initialize
	  // the regeneratorRuntime variable at the top of this file.
	  module.exports
	));

	try {
	  regeneratorRuntime = runtime;
	} catch (accidentalStrictMode) {
	  // This module should not be running in strict mode, so the above
	  // assignment should always work unless something is misconfigured. Just
	  // in case runtime.js accidentally runs in strict mode, we can escape
	  // strict mode using a global Function call. This could conceivably fail
	  // if a Content Security Policy forbids using Function, but in that case
	  // the proper solution is to fix the accidental strict mode problem. If
	  // you've misconfigured your bundler to force strict mode and applied a
	  // CSP to forbid Function, and you're not willing to fix either of those
	  // problems, please detail your unique predicament in a GitHub issue.
	  Function("r", "regeneratorRuntime = r")(runtime);
	}
	});

	var regenerator = runtime_1;

	function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
	  try {
	    var info = gen[key](arg);
	    var value = info.value;
	  } catch (error) {
	    reject(error);
	    return;
	  }

	  if (info.done) {
	    resolve(value);
	  } else {
	    Promise.resolve(value).then(_next, _throw);
	  }
	}

	function _asyncToGenerator(fn) {
	  return function () {
	    var self = this,
	        args = arguments;
	    return new Promise(function (resolve, reject) {
	      var gen = fn.apply(self, args);

	      function _next(value) {
	        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
	      }

	      function _throw(err) {
	        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
	      }

	      _next(undefined);
	    });
	  };
	}

	var asyncToGenerator = _asyncToGenerator;

	function _classCallCheck(instance, Constructor) {
	  if (!(instance instanceof Constructor)) {
	    throw new TypeError("Cannot call a class as a function");
	  }
	}

	var classCallCheck = _classCallCheck;

	function _defineProperties(target, props) {
	  for (var i = 0; i < props.length; i++) {
	    var descriptor = props[i];
	    descriptor.enumerable = descriptor.enumerable || false;
	    descriptor.configurable = true;
	    if ("value" in descriptor) descriptor.writable = true;
	    Object.defineProperty(target, descriptor.key, descriptor);
	  }
	}

	function _createClass(Constructor, protoProps, staticProps) {
	  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
	  if (staticProps) _defineProperties(Constructor, staticProps);
	  return Constructor;
	}

	var createClass = _createClass;

	var _typeof_1 = createCommonjsModule(function (module) {
	function _typeof2(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof2 = function _typeof2(obj) { return typeof obj; }; } else { _typeof2 = function _typeof2(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof2(obj); }

	function _typeof(obj) {
	  if (typeof Symbol === "function" && _typeof2(Symbol.iterator) === "symbol") {
	    module.exports = _typeof = function _typeof(obj) {
	      return _typeof2(obj);
	    };
	  } else {
	    module.exports = _typeof = function _typeof(obj) {
	      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : _typeof2(obj);
	    };
	  }

	  return _typeof(obj);
	}

	module.exports = _typeof;
	});

	function _assertThisInitialized(self) {
	  if (self === void 0) {
	    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
	  }

	  return self;
	}

	var assertThisInitialized = _assertThisInitialized;

	function _possibleConstructorReturn(self, call) {
	  if (call && (_typeof_1(call) === "object" || typeof call === "function")) {
	    return call;
	  }

	  return assertThisInitialized(self);
	}

	var possibleConstructorReturn = _possibleConstructorReturn;

	var getPrototypeOf = createCommonjsModule(function (module) {
	function _getPrototypeOf(o) {
	  module.exports = _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
	    return o.__proto__ || Object.getPrototypeOf(o);
	  };
	  return _getPrototypeOf(o);
	}

	module.exports = _getPrototypeOf;
	});

	var setPrototypeOf = createCommonjsModule(function (module) {
	function _setPrototypeOf(o, p) {
	  module.exports = _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
	    o.__proto__ = p;
	    return o;
	  };

	  return _setPrototypeOf(o, p);
	}

	module.exports = _setPrototypeOf;
	});

	function _inherits(subClass, superClass) {
	  if (typeof superClass !== "function" && superClass !== null) {
	    throw new TypeError("Super expression must either be null or a function");
	  }

	  subClass.prototype = Object.create(superClass && superClass.prototype, {
	    constructor: {
	      value: subClass,
	      writable: true,
	      configurable: true
	    }
	  });
	  if (superClass) setPrototypeOf(subClass, superClass);
	}

	var inherits = _inherits;

	var Popup =
	/*#__PURE__*/
	function (_Handler) {
	  inherits(Popup, _Handler);

	  function Popup(config) {
	    var _this;

	    classCallCheck(this, Popup);

	    _this = possibleConstructorReturn(this, getPrototypeOf(Popup).call(this, config));
	    _this.config = salteAuth_min_6.Common.defaults(_this.config, {
	      window: {
	        name: '@salte-auth/tab',
	        height: 600,
	        width: 600
	      }
	    });
	    return _this;
	  }

	  createClass(Popup, [{
	    key: "open",
	    value: function () {
	      var _open = asyncToGenerator(
	      /*#__PURE__*/
	      regenerator.mark(function _callee(options) {
	        var top, left, popupWindow;
	        return regenerator.wrap(function _callee$(_context) {
	          while (1) {
	            switch (_context.prev = _context.next) {
	              case 0:
	                top = window.innerHeight / 2 - this.config.window.height / 2 + window.screenTop;
	                left = window.innerWidth / 2 - this.config.window.width / 2 + window.screenLeft;
	                popupWindow = window.open(options.url, this.config.window.name, "height=".concat(this.config.window.height, ", width=").concat(this.config.window.width, ", status=yes, toolbar=no, menubar=no, location=no, top=").concat(top, ", left=").concat(left));

	                if (popupWindow) {
	                  _context.next = 5;
	                  break;
	                }

	                throw new salteAuth_min_2({
	                  message: 'We were unable to open the popup window, its likely that the request was blocked.',
	                  code: 'popup_blocked'
	                });

	              case 5:
	                popupWindow.focus(); // TODO: Find a better way of tracking when a Window closes.

	                return _context.abrupt("return", new Promise(function (resolve, reject) {
	                  var checker = setInterval(function () {
	                    try {
	                      if (!popupWindow.closed) {
	                        // This could throw cross-domain errors, so we need to silence them.
	                        if (popupWindow.location.href.indexOf(options.redirectUrl) !== 0) return;
	                        var parsed = salteAuth_min_6.URL.parse(popupWindow.location);
	                        popupWindow.close();
	                        resolve(parsed);
	                      }

	                      clearInterval(checker);
	                    } catch (error) {
	                      if (salteAuth_min_6.Events.isCrossDomainError(error)) return;
	                      popupWindow.close();
	                      clearInterval(checker);
	                      reject(error);
	                    }
	                  }, 100);
	                }));

	              case 7:
	              case "end":
	                return _context.stop();
	            }
	          }
	        }, _callee, this);
	      }));

	      function open(_x) {
	        return _open.apply(this, arguments);
	      }

	      return open;
	    }()
	  }, {
	    key: "name",
	    get: function get() {
	      return 'popup';
	    }
	  }, {
	    key: "auto",
	    get: function get() {
	      return false;
	    }
	  }]);

	  return Popup;
	}(salteAuth_min_5);

	var auth = new salteAuth_min_1({
	  providers: [new salteAuth_min_7.OAuth2({
	    login: function login() {
	      return 'https://github.com/login/oauth/authorize';
	    },
	    clientID: 'b44780ca7678681180c9',
	    responseType: 'code',
	    routes: true
	  })],
	  handlers: [new Popup({
	    default: true
	  })]
	});
	auth.on('login', function (error, idToken) {
	  if (error) console.error(error);else console.log(idToken);
	});
	var button = document.createElement('button');
	button.innerHTML = "Login";
	button.addEventListener('click', function () {
	  auth.login('generic.oauth2');
	});
	document.body.appendChild(button);

}));
