// @kepler.gl/components/dist/common/progress-bar.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject, _templateObject2;

var StyledBar = _styledComponents["default"].span.attrs({
  className: 'progress-bar__bar'
})(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n  /* transition: width 200ms; */\n  display: block;\n"])), function (props) {
  return props.barColor || props.theme.progressBarColor;
});

var StyledTrack = _styledComponents["default"].div.attrs({
  className: 'progress-bar'
})(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n"])), function (props) {
  return props.trackColor || props.theme.progressBarTrackColor;
});

/** @type {React.FunctionComponent<ProgressBarProps>} */
var ProgressBar = function ProgressBar(_ref) {
  var percent = _ref.percent,
      _ref$height = _ref.height,
      height = _ref$height === void 0 ? 4 : _ref$height,
      isLoading = _ref.isLoading,
      barColor = _ref.barColor,
      trackColor = _ref.trackColor,
      theme = _ref.theme;
  return /*#__PURE__*/_react["default"].createElement(StyledTrack, {
    trackColor: trackColor,
    theme: theme
  }, /*#__PURE__*/_react["default"].createElement(StyledBar, {
    barColor: barColor,
    style: {
      width: percent,
      height: "".concat(height, "px"),
      opacity: isLoading ? 1 : 0
    }
  }));
};

var _default = ProgressBar;
exports["default"] = _default;
