// @loaders.gl/json/dist/ndjson-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NDJSONLoader = void 0;
const parse_ndjson_1 = __importDefault(require("./lib/parse-ndjson"));
const parse_ndjson_in_batches_1 = __importDefault(require("./lib/parse-ndjson-in-batches"));
// __VERSION__ is injected by babel-plugin-version-inline
// @ts-ignore TS2304: Cannot find name '__VERSION__'.
const VERSION = typeof __VERSION__ !== 'undefined' ? __VERSION__ : 'latest';
exports.NDJSONLoader = {
    name: 'NDJSON',
    id: 'ndjson',
    module: 'json',
    version: VERSION,
    extensions: ['ndjson', 'jsonl'],
    mimeTypes: [
        'application/x-ndjson',
        'application/jsonlines',
        'application/json-seq'
    ],
    category: 'table',
    text: true,
    parse: async (arrayBuffer) => (0, parse_ndjson_1.default)(new TextDecoder().decode(arrayBuffer)),
    parseTextSync: parse_ndjson_1.default,
    parseInBatches: parse_ndjson_in_batches_1.default,
    options: {}
};
