// @loaders.gl/wms/dist/index.js
"use strict";
// loaders.gl, MIT license
Object.defineProperty(exports, "__esModule", { value: true });
exports._GMLLoader = exports._WFSCapabilitiesLoader = exports._WMTSCapabilitiesLoader = exports._WMSLayerDescriptionLoader = exports._WMSFeatureInfoLoader = exports._ArcGISImageService = exports.WMSService = exports.ImageService = exports.createImageSource = exports.ImageSource = exports.WMSCapabilitiesLoader = exports.WMSErrorLoader = void 0;
// WMS
var wms_error_loader_1 = require("./wms-error-loader");
Object.defineProperty(exports, "WMSErrorLoader", { enumerable: true, get: function () { return wms_error_loader_1.WMSErrorLoader; } });
var wms_capabilities_loader_1 = require("./wms-capabilities-loader");
Object.defineProperty(exports, "WMSCapabilitiesLoader", { enumerable: true, get: function () { return wms_capabilities_loader_1.WMSCapabilitiesLoader; } });
var image_source_1 = require("./lib/data-sources/image-source");
Object.defineProperty(exports, "ImageSource", { enumerable: true, get: function () { return image_source_1.ImageSource; } });
var create_image_source_1 = require("./lib/data-sources/create-image-source");
Object.defineProperty(exports, "createImageSource", { enumerable: true, get: function () { return create_image_source_1.createImageSource; } });
var image_service_1 = require("./lib/data-sources/image-services/image-service");
Object.defineProperty(exports, "ImageService", { enumerable: true, get: function () { return image_service_1.ImageService; } });
var wms_service_1 = require("./lib/data-sources/image-services/wms-service");
Object.defineProperty(exports, "WMSService", { enumerable: true, get: function () { return wms_service_1.WMSService; } });
var arcgis_image_service_1 = require("./lib/data-sources/image-services/arcgis-image-service");
Object.defineProperty(exports, "_ArcGISImageService", { enumerable: true, get: function () { return arcgis_image_service_1.ArcGISImageService; } });
var wms_feature_info_loader_1 = require("./wip/wms-feature-info-loader");
Object.defineProperty(exports, "_WMSFeatureInfoLoader", { enumerable: true, get: function () { return wms_feature_info_loader_1.WMSFeatureInfoLoader; } });
var wms_layer_description_loader_1 = require("./wip/wms-layer-description-loader");
Object.defineProperty(exports, "_WMSLayerDescriptionLoader", { enumerable: true, get: function () { return wms_layer_description_loader_1.WMSLayerDescriptionLoader; } });
var wmts_capabilities_loader_1 = require("./wip/wmts-capabilities-loader");
Object.defineProperty(exports, "_WMTSCapabilitiesLoader", { enumerable: true, get: function () { return wmts_capabilities_loader_1.WMTSCapabilitiesLoader; } });
var wfs_capabilities_loader_1 = require("./wip/wfs-capabilities-loader");
Object.defineProperty(exports, "_WFSCapabilitiesLoader", { enumerable: true, get: function () { return wfs_capabilities_loader_1.WFSCapabilitiesLoader; } });
var gml_loader_1 = require("./wip/gml-loader");
Object.defineProperty(exports, "_GMLLoader", { enumerable: true, get: function () { return gml_loader_1.GMLLoader; } });
// LERC
// TODO - restore once esbuild bundling issues have been resolved
// export type {LERCData} from './lib/lerc/lerc-types.js';
// export {LERCLoader} from './lerc-loader.js';
