// @loaders.gl/wms/dist/lib/data-sources/image-services/image-service.js
"use strict";
// loaders.gl, MIT license
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeImageServiceProps = exports.getFetchFunction = exports.ImageService = void 0;
const images_1 = require("@loaders.gl/images");
const image_source_1 = require("../image-source");
/**
 * Quickly connect to "ad hoc" image sources without subclassing ImageSource.
 * ImageSource allows template url strings to be used to ad hoc connect to arbitrary image data sources
 * Accepts a template url string and builds requests URLs
 */
class ImageService extends image_source_1.ImageSource {
    constructor(props) {
        super();
        this.props = mergeImageServiceProps(props);
        this.fetch = getFetchFunction(props);
    }
    // IMAGE SOURCE API
    async getMetadata() {
        throw new Error('ImageSource.getMetadata not implemented');
    }
    async getImage(parameters) {
        const granularParameters = this.getGranularParameters(parameters);
        const url = this.getURLFromTemplate(granularParameters);
        const response = await this.fetch(url);
        const arrayBuffer = await response.arrayBuffer();
        return await images_1.ImageLoader.parse(arrayBuffer);
    }
    // HELPERS
    /** Break up bounding box in east, north, south, west */
    getGranularParameters(parameters) {
        const [east, north, west, south] = parameters.bbox;
        return { ...parameters, east, north, south, west };
    }
    /** Supports both ${} and {} notations */
    getURLFromTemplate(parameters) {
        let url = this.props.url;
        for (const [key, value] of Object.entries(parameters)) {
            // TODO - parameter could be repeated
            // const regex = new RegExp(`\${${key}}`, 'g');
            url = url.replace(`\${${key}}`, String(value));
            url = url.replace(`{${key}}`, String(value));
        }
        return url;
    }
}
exports.ImageService = ImageService;
ImageService.type = 'template';
ImageService.testURL = (url) => url.toLowerCase().includes('{');
/**
 * Gets the current fetch function from options
 * @todo - move to loader-utils module
 * @todo - use in core module counterpart
 * @param options
 * @param context
 */
function getFetchFunction(options) {
    const fetchFunction = options?.fetch;
    // options.fetch can be a function
    if (fetchFunction && typeof fetchFunction === 'function') {
        return (url, fetchOptions) => fetchFunction(url, fetchOptions);
    }
    // options.fetch can be an options object, use global fetch with those options
    const fetchOptions = options?.fetch;
    if (fetchOptions && typeof fetchOptions !== 'function') {
        return (url) => fetch(url, fetchOptions);
    }
    // else return the global fetch function
    return (url) => fetch(url);
}
exports.getFetchFunction = getFetchFunction;
function mergeImageServiceProps(props) {
    return {
        // Default fetch
        ...props,
        loadOptions: {
            ...props.loadOptions,
            fetch: getFetchFunction(props.loadOptions)
        }
    };
}
exports.mergeImageServiceProps = mergeImageServiceProps;
