// react-map-gl-draw/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ELEMENT_TYPE = exports.STATIC_STYLE = exports.OPERATIONS = exports.RENDER_STATE = exports.MODE_TO_RENDER_TYPE = exports.RENDER_TYPE = exports.MODE_TO_GEOJSON_TYPE = exports.GEOJSON_TYPE = exports.DRAWING_MODES = exports.MODES = void 0;

var _MODE_TO_GEOJSON_TYPE, _MODE_TO_RENDER_TYPE;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var MODES = {
  READ_ONLY: 'READ_ONLY',
  SELECT_FEATURE: 'SELECT_FEATURE',
  EDIT_VERTEX: 'EDIT_VERTEX',
  DRAW_POINT: 'DRAW_POINT',
  DRAW_PATH: 'DRAW_PATH',
  DRAW_POLYGON: 'DRAW_POLYGON',
  DRAW_RECTANGLE: 'DRAW_RECTANGLE'
};
exports.MODES = MODES;
var DRAWING_MODES = [MODES.DRAW_POINT, MODES.DRAW_PATH, MODES.DRAW_POLYGON, MODES.DRAW_RECTANGLE];
exports.DRAWING_MODES = DRAWING_MODES;
var GEOJSON_TYPE = {
  POINT: 'Point',
  LINE_STRING: 'LineString',
  POLYGON: 'Polygon'
};
exports.GEOJSON_TYPE = GEOJSON_TYPE;
var MODE_TO_GEOJSON_TYPE = (_MODE_TO_GEOJSON_TYPE = {}, _defineProperty(_MODE_TO_GEOJSON_TYPE, MODES.DRAW_POINT, GEOJSON_TYPE.POINT), _defineProperty(_MODE_TO_GEOJSON_TYPE, MODES.DRAW_PATH, GEOJSON_TYPE.LINE_STRING), _defineProperty(_MODE_TO_GEOJSON_TYPE, MODES.DRAW_POLYGON, GEOJSON_TYPE.POLYGON), _defineProperty(_MODE_TO_GEOJSON_TYPE, MODES.DRAW_RECTANGLE, GEOJSON_TYPE.POLYGON), _MODE_TO_GEOJSON_TYPE);
exports.MODE_TO_GEOJSON_TYPE = MODE_TO_GEOJSON_TYPE;
var RENDER_TYPE = {
  POINT: 'Point',
  LINE_STRING: 'LineString',
  POLYGON: 'Polygon',
  RECTANGLE: 'Rectangle'
};
exports.RENDER_TYPE = RENDER_TYPE;
var MODE_TO_RENDER_TYPE = (_MODE_TO_RENDER_TYPE = {}, _defineProperty(_MODE_TO_RENDER_TYPE, MODES.DRAW_POINT, RENDER_TYPE.POINT), _defineProperty(_MODE_TO_RENDER_TYPE, MODES.DRAW_PATH, RENDER_TYPE.LINE_STRING), _defineProperty(_MODE_TO_RENDER_TYPE, MODES.DRAW_POLYGON, RENDER_TYPE.POLYGON), _defineProperty(_MODE_TO_RENDER_TYPE, MODES.DRAW_RECTANGLE, RENDER_TYPE.RECTANGLE), _MODE_TO_RENDER_TYPE);
exports.MODE_TO_RENDER_TYPE = MODE_TO_RENDER_TYPE;
var RENDER_STATE = {
  INACTIVE: 'INACTIVE',
  UNCOMMITTED: 'UNCOMMITTED',
  SELECTED: 'SELECTED',
  HOVERED: 'HOVERED',
  CLOSING: 'CLOSING'
};
exports.RENDER_STATE = RENDER_STATE;
var OPERATIONS = {
  NONE: 'NONE',
  SET: 'SET',
  INTERSECT: 'INTERSECT',
  INSERT: 'INSERT'
};
exports.OPERATIONS = OPERATIONS;
var STATIC_STYLE = {
  cursor: 'default',
  pointerEvents: 'none'
};
exports.STATIC_STYLE = STATIC_STYLE;
var ELEMENT_TYPE = {
  FEATURE: 'feature',
  SEGMENT: 'segment',
  VERTEX: 'vertex'
};
exports.ELEMENT_TYPE = ELEMENT_TYPE;
