// react-map-gl-draw/dist/feature.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _constants = require("./constants");

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Feature =
/*#__PURE__*/
function () {
  _createClass(Feature, null, [{
    key: "fromFeature",
    value: function fromFeature(feature) {
      var id = feature.id,
          _feature$geometry = feature.geometry,
          coordinates = _feature$geometry.coordinates,
          type = _feature$geometry.type,
          _feature$properties = feature.properties,
          renderType = _feature$properties.renderType,
          otherProps = _objectWithoutProperties(_feature$properties, ["renderType"]);

      switch (type) {
        case _constants.GEOJSON_TYPE.POINT:
          return new Feature({
            id: id,
            type: type,
            renderType: renderType,
            points: [coordinates],
            otherProps: otherProps
          });

        case _constants.GEOJSON_TYPE.LINE_STRING:
          return new Feature({
            id: id,
            type: type,
            renderType: renderType,
            points: coordinates,
            otherProps: otherProps
          });

        case _constants.GEOJSON_TYPE.POLYGON:
          var points = coordinates[0] && coordinates[0].slice(0, -1);
          return new Feature({
            id: id,
            type: type,
            renderType: renderType,
            points: points,
            isClosed: true,
            otherProps: otherProps
          });

        default:
          return null;
      }
    }
  }]);

  function Feature(props) {
    _classCallCheck(this, Feature);

    _defineProperty(this, "id", void 0);

    _defineProperty(this, "type", void 0);

    _defineProperty(this, "renderType", void 0);

    _defineProperty(this, "isClosed", false);

    _defineProperty(this, "points", void 0);

    _defineProperty(this, "otherProps", void 0);

    this.id = props.id;
    this.type = props.type;
    this.renderType = props.renderType;
    this.points = props.points || [];
    this.isClosed = props.isClosed;
    this.otherProps = props.otherProps;
  }

  _createClass(Feature, [{
    key: "addPoint",
    value: function addPoint(pt) {
      this.points.push(pt);
      return true;
    }
  }, {
    key: "insertPoint",
    value: function insertPoint(pt, index) {
      this.points.splice(index, 0, pt);
    }
  }, {
    key: "getBoundingBox",
    value: function getBoundingBox() {
      if (!this.points || !this.points.length) {
        return null;
      }

      var bbox = this.points.reduce(function (result, pt) {
        result.xmin = Math.min(pt[0], result.xmin);
        result.xmax = Math.min(pt[0], result.xmax);
        result.ymin = Math.min(pt[1], result.ymin);
        result.ymax = Math.min(pt[1], result.ymax);
        return result;
      }, {
        xmin: Infinity,
        xmax: -Infinity,
        ymin: Infinity,
        ymax: -Infinity
      });
      return bbox;
    }
  }, {
    key: "removePoint",
    value: function removePoint(index) {
      var points = this.points;

      if (index >= 0 && index < points.length) {
        points.splice(index, 1);

        if (points.length < 3) {
          this.isClosed = false;
        }

        return true;
      }

      return false;
    }
  }, {
    key: "replacePoint",
    value: function replacePoint(index, pt) {
      var points = this.points;

      if (index >= 0 && index < points.length) {
        points[index] = pt;
        return true;
      }

      return false;
    }
  }, {
    key: "closePath",
    value: function closePath() {
      var points = this.points;

      if (points.length >= 3 && !this.isClosed) {
        this.isClosed = true;
        return true;
      }

      return false;
    }
  }, {
    key: "toFeature",
    value: function toFeature() {
      var id = this.id,
          points = this.points,
          isClosed = this.isClosed,
          renderType = this.renderType,
          otherProps = this.otherProps;
      var feature = null;

      if (points.length < 2) {
        feature = {
          type: 'Feature',
          geometry: {
            type: _constants.GEOJSON_TYPE.POINT,
            coordinates: points[0]
          },
          properties: _objectSpread({
            renderType: renderType
          }, otherProps),
          id: id
        };
      } else if (points.length < 3 || !isClosed) {
        feature = {
          type: 'Feature',
          geometry: {
            type: _constants.GEOJSON_TYPE.LINE_STRING,
            coordinates: points
          },
          properties: _objectSpread({
            renderType: renderType,
            bbox: this.getBoundingBox()
          }, otherProps),
          id: id
        };
      } else {
        feature = {
          type: 'Feature',
          geometry: {
            type: _constants.GEOJSON_TYPE.POLYGON,
            coordinates: [_toConsumableArray(points).concat([points[0]])]
          },
          properties: _objectSpread({
            renderType: renderType,
            isClosed: isClosed,
            bbox: this.getBoundingBox()
          }, otherProps),
          id: id
        };
      }

      return feature;
    }
  }]);

  return Feature;
}();

exports.default = Feature;
