// react-map-gl-draw/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findClosestPointOnLineSegment = findClosestPointOnLineSegment;
exports.isNumeric = isNumeric;
exports.parseElemDataAttributes = parseElemDataAttributes;

function inBounds(p1, p2, p) {
  var bounds = [Math.min(p1[0], p2[0]), Math.max(p1[0], p2[0]), Math.min(p1[1], p2[1]), Math.max(p1[1], p2[1])];
  return p[0] >= bounds[0] && p[0] <= bounds[1] && p[1] >= bounds[2] && p[1] <= bounds[3];
}

function findClosestPointOnLineSegment(p1, p2, p) {
  // line
  var k = (p2[1] - p1[1]) / (p2[0] - p1[0]);
  var b = p1[1] - k * p1[0]; // vertical line

  if (!isFinite(k)) {
    var q = [p1[0], p[1]];
    return inBounds(p1, p2, q) ? q : null;
  } // p is on line [p1, p2]


  if (p[0] * k + b - p[1] === 0) {
    return inBounds(p1, p2, p) ? p : null;
  }

  var qx = (k * p[1] + p[0] - k * b) / (k * k + 1);
  var qy = k * qx + b;
  return inBounds(p1, p2, [qx, qy]) ? [qx, qy] : null;
}

function isNumeric(n) {
  return !Array.isArray(n) && !isNaN(parseFloat(n)) && isFinite(n);
}

function parseElemDataAttributes(elem) {
  var data = elem && elem.dataset;

  if (!data) {
    return null;
  }

  var featureIndex = data.featureIndex;
  var vertexIndex = data.vertexIndex;
  return {
    type: data.type,
    operation: data.operation,
    featureIndex: isNumeric(featureIndex) ? Number(featureIndex) : undefined,
    vertexIndex: isNumeric(vertexIndex) ? Number(vertexIndex) : undefined
  };
}
