// @kepler.gl/constants/dist/user-guides.js
// Copyright (c) 2022 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BUG_REPORT_LINK = exports.GUIDES_FILE_FORMAT_DOC = exports.EXPORT_HTML_MAP_MODES_DOC = exports.MAPBOX_ACCESS_TOKEN = exports.ADD_DATA_TO_MAP_DOC = exports.EXPORT_HTML_MAP_DOC = exports.API_REFERENCE_DOC = exports.ACTIONS_DOC = exports.USER_GUIDE_DOC = void 0;var USER_GUIDE_DOC = 'https://docs.kepler.gl/docs/user-guides';
exports.USER_GUIDE_DOC = USER_GUIDE_DOC;
var ACTIONS_DOC = 'https://docs.kepler.gl/docs/api-reference/actions/actions';
exports.ACTIONS_DOC = ACTIONS_DOC;
var API_REFERENCE_DOC = 'https://docs.kepler.gl/docs/api-reference';
exports.API_REFERENCE_DOC = API_REFERENCE_DOC;
var EXPORT_HTML_MAP_DOC = "".concat(USER_GUIDE_DOC, "/k-save-and-export");
exports.EXPORT_HTML_MAP_DOC = EXPORT_HTML_MAP_DOC;
var ADD_DATA_TO_MAP_DOC = "".concat(ACTIONS_DOC, "#adddatatomap");
exports.ADD_DATA_TO_MAP_DOC = ADD_DATA_TO_MAP_DOC;
var MAPBOX_ACCESS_TOKEN = 'https://docs.mapbox.com/help/how-mapbox-works/access-tokens/';
exports.MAPBOX_ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;
var EXPORT_HTML_MAP_MODES_DOC = "".concat(USER_GUIDE_DOC, "/k-save-and-export#export-html-map");
exports.EXPORT_HTML_MAP_MODES_DOC = EXPORT_HTML_MAP_MODES_DOC;
var GUIDES_FILE_FORMAT_DOC = "".concat(USER_GUIDE_DOC, "/b-kepler-gl-workflow/a-add-data-to-the-map#supported-file-formats");
exports.GUIDES_FILE_FORMAT_DOC = GUIDES_FILE_FORMAT_DOC;
var BUG_REPORT_LINK = 'https://github.com/keplergl/kepler.gl/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBug%5D';
exports.BUG_REPORT_LINK = BUG_REPORT_LINK;
